<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<HTML>
<HEAD>
   <META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
   <META name="keywords" content="BLASTA, WU-BLAST, WU BLAST, WUBLAST">
   <TITLE>WU&nbsp;BLAST 2.0 Description</TITLE>
</HEAD>
<BODY>

<H2>WU&nbsp;BLAST&nbsp;2.0 TOPICS</H2>
<P>
Copyright (c) 2003 Warren R. Gish, Saint Louis, Missouri 63108 USA.
</P>
<P>
All Rights Reserved.
</P>
<HR>
<UL>
<LI><A HREF="#Description">Description</A></LI>
<LI><A HREF="#Features">Key Features</A></LI>
<LI><A HREF="#Manifest">Manifest</A></LI>
<LI><A HREF="#Tofly">To Fly...</A> (last updated <B>2001-11-14</B>)</LI>
<LI><A HREF="http://blast.wustl.edu/blast/cparms.html">Comparable WU/NCBI&nbsp;BLAST Parameters</A></LI>
<LI><A HREF="#Examples">Examples</A></LI>
<LI><A HREF="#Options">New Command Line Options</A></LI>
<LI><A HREF="#EnvVars">Environment Variables</A></LI>
<LI><A HREF="#Filters">Filters and Masks</A></LI>
<LI><A HREF="#Bugs">Bugs</A></LI>
<LI><A HREF="Memory.html" TARGET="_blank">Memory Requirements</A></LI>
<LI><A HREF="#Platforms">Supported Platforms</A></LI>
<LI><A HREF="http://blast.wustl.edu/blast/executables" TARGET="_blank">Precompiled Executables (Old and Obsolete but Free)</A></LI>
<LI><A HREF="#Installation">Installation</A></LI>
<LI><A HREF="#Licensing">Licensing of the Current Version 2.0 Software</A></LI>
<LI><A HREF="#Citing">Citing BLAST</A></LI>
<LI><A HREF="#History">Historical Notes</A></LI>
<LI><A HREF="#References">References</A></LI>
</UL>

<HR>
<A NAME="Description"></A>
<H3>
Description</H3>
<P>
Washington University BLAST (<B>WU&nbsp;BLAST</B>) version 2.0 is a powerful
software package for gene and protein identification,
using sensitive, selective and rapid similarity searches of
protein and nucleotide sequence databases.
WU&nbsp;BLAST&nbsp;2.0 builds upon
<A HREF="http://blast.wustl.edu/blast-1.4">WU&nbsp;BLAST&nbsp;1.4</A>,
which in turn was based on the public domain
NCBI&nbsp;BLAST version 1.4
(Gish, unpublished, 1994;
<A HREF="#Altschul_et_al_1990">Altschul <I>et al.</I>, 1990</A>;
<A HREF="#Gish_and_States_1993">Gish and States, 1993</A>).
While NCBI&nbsp;BLAST and WU&nbsp;BLAST&nbsp;1.4 are in the public domain,
WU&nbsp;BLAST&nbsp;2.0 contains significant new features and extended capabilities,
the development of which began in late 1994, at Washington University in Saint Louis.
First released in May 1996, or more than a year ahead of the NCBI,
WU&nbsp;BLAST&nbsp;2.0 is the original gapped BLAST with statistics and
is known for setting higher standards for sensitivity, speed, correctness and accuracy,
scalability and reliability than competing programs and implementations.
WU&nbsp;BLAST is not a re-hash of NCBI&nbsp;BLAST and essentially shares
no code with it, except for small portions that both packages
derived from ungapped NCBI&nbsp;BLAST&nbsp;1.4.
</P>
<P>
<B>
WU&nbsp;BLAST has been built to be the most trusted database search tool
in your software arsenal.
Its unique combination of speed, accuracy, efficiency, flexibility,
scalability, reliability
and consistency across all supported platforms
is achieved through careful software coding,
the use of extensive error checks,
anticipation of future needs, and superior design.
</B>
</P>

<P>
[<B>Note</B>:&nbsp;&nbsp;
In spite of many similar or identical characteristics of the algorithms employed,
WU&nbsp;BLAST&nbsp;2.0 and NCBI&nbsp;Gapped BLAST are distinctly
different software packages that, in ways of varying importance, carry out their work differently.
Consequently, the two packages often yield different results, particularly in the areas of
the default level of sensitivity, details in how the statistics are employed,
and occasionally in the accuracy or completeness of the results.]
</P>

<P>
The feature list for the licensed version of WU&nbsp;BLAST&nbsp;2.0 is large
and continues to expand.
Much of this is outlined
<A HREF="http://blast.wustl.edu/blast/README.html#Features">below</A>.
The primary purpose of the freely available version 2.0a19
is to allow users to demonstrate for
themselves the effectiveness of using gapped alignments instead of ungapped,
when combined with the evaluation of the joint probability of multiple regions of similarity,
using
<A HREF="#Karlin_and_Altschul_1993">Karlin and Altschul (1993)</A> "Sum" statistics.
Not surprisingly,
with version 2.0a19
one can obtain markedly improved
results over
<A HREF="http://blast.wustl.edu/blast-1.4">version&nbsp;1.4,</A>
primarily due to the introduction of gapped alignments.
WU&nbsp;BLAST&nbsp;2.0a19 executables for several UNIX platforms can be downloaded
from <A HREF="http://blast.wustl.edu/blast/executables">http://blast.wustl.edu/blast/executables</A>.
The complete suite of search programs (blastp, blastn, blastx, tblastn,
and tblastx) is included, as well as several support programs.
Users of the freely available version 2.0a19 should keep in mind
that its reliability, features, flexibility, scalability and speed
are generally not comparable to the licensed version 2.0.
</P>

<P>
WU&nbsp;BLAST&nbsp;2.0 is <A HREF="http://blast.wustl.edu/blast/Copyright">copyrighted</A>
and may not be sold, redistributed or modified in any form or by any means,
without the express written consent of the Washington University School of Medicine in St. Louis.
Other than the aforementioned restrictions,
the version 2.0a19 executables posted
<A HREF="http://blast.wustl.edu/blast/executables" TARGET="_blank">here</A>
may be freely used for commercial, nonprofit, or academic purposes.
</P>

<P>
DISCLAIMER: THIS SOFTWARE IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND.
</P>

<HR>
<A NAME="Features"></A>
<H3>Key Features</H3>

<P>
Some of the key features of WU&nbsp;BLAST&nbsp;2.0 are described below,
many of which are only available in the licensed version.
</P>

<UL>
<LI>
Gapped alignment routines are available (and used by default)
in <I>all</I> BLAST search modes:
<TT>BLASTP</TT>, <TT>BLASTN</TT>,
and <TT>TBLASTN</TT> (<A HREF="#Altschul_et_al_1990">Altschul <I>et&nbsp;al.</I>,&nbsp;1990</A>),
as well as, <TT>BLASTX</TT> (<A HREF="#Gish_and_States_1993">Gish and States,&nbsp;1993</A>)
and <TT>TBLASTX</TT> (Gish, W., 1994, unpublished).
Gaps can optionally be turned off in any mode if desired.
</LI>

<LI>
Potentially <I>multiple</I> regions
of similarity are identified and reported for each database sequence,
thus yielding increased sensitivity and selectivity.
This feature is essential for finding:
all exons in a multi-exon gene sequence,
not just the longest or best-matching exon;
all complete or partial copies of a repetitive element
in a genomic sequence, not just the best matching one;
and multiple, discrete domains of similarity between sequences,
not just the highest-scoring one.
</LI>

<LI>
<A HREF="#Karlin_and_Altschul_1993">Karlin and Altschul (1993)</A> "Sum statistics" are
available (and used by default)
in <I>all</I> search modes,
to evaluate the <I>joint</I> probability of multiple regions of similarity,
as described by <A HREF="#Altschul_and_Gish_1996">Altschul and Gish (1996)</A>.
By this technique, sets of similar regions are often found to be statistically significant
that individually would be <I>in</I>significant and go unreported.
The combination of well-chosen heuristics and statistics in WU&nbsp;BLAST is often
more sensitive/selective than:
the full dynamic programming approach of
<A HREF="#Smith_Waterman_1981">
Smith and Waterman (1981)</A>,
that finds and evaluates the significance
of only the highest scoring alignment with each database sequence;
and other approaches or BLAST implementations that
identify multiple regions of local similarity which are then evaluated individually
for statistical significance rather than jointly.
</LI>

<LI>
Poisson statistics are available as an option to Karlin-Altschul Sum statistics
in all search modes.
Simpler
<A HREF="#Karlin_and_Altschul_1990">Karlin-Altschul (1990)</A>
statistics,
that do not involve joint probability calculations,
are also available as an option.
</LI>

<LI>
Using the <B>postsw</B> option,
a full Smith-Waterman alignment is performed on query-subject pairs
of sequences that will be reported by <TT>BLASTP</TT>.
The Smith-Waterman scores and alignments are combined with
the initial BLAST results and redundancy is removed.
This may alter the relative ranking of database matches
before output.
Use of this option is recommended,
although it may be supplanted in the future by other option(s) or by a redefined default behavior.
</LI>

<LI>
The execution of WU&nbsp;BLAST&nbsp;2.0 has been optimized
such that gapped searches typically run <I>faster</I>
than the ungapped version 1.4 programs, when using identical ungapped BLAST parameters.
An exception to this rule is <TT>BLASTN</TT> (already quite fast) which runs about 10% slower
with the new gapped alignment searches and default parameters.
Any of the search modes may still run more slowly,
however, if non-random sequence
such as poly-A tracts, repetitive elements, and low-complexity regions
are not filtered or masked from the query or database prior
to performing the search.
</LI>

<LI>
The classical ungapped BLAST algorithm has not been changed in WU&nbsp;BLAST&nbsp;2.0,
thus retaining the sensitivity and control characteristics
that users became accustomed to with previous versions of BLAST.
When assessed at the same sensitivity level,
the optimized, classical BLAST algorithm in WU&nbsp;BLAST&nbsp;2.0 is nearly the same speed
and uses much less memory than the 2-hit algorithm described by
<A HREF="#Altschul_et_al_1997">Altschul <I>et al.</I> (1997)</A>.
For users who wish even higher speed,
a 2-hit algorithm is available
in a higher-performance WU implementation (see the <B>hitdist</B> option).
While the classical 1-hit BLAST algorithm is the default in all WU&nbsp;BLAST search modes,
the 2-hit algorithm is available as an option in all search modes,
including <TT>BLASTN</TT>.
</LI>

<LI>
WU&nbsp;BLAST&nbsp;2.0 is a virtual drop-in replacement for version 1.4,
utilizing the same inputs and command line arguments,
while producing almost the same format of output as before.
A parser of version 1.4 output is only prone to breaking on version 2 output
if it checks for just a few kinds of consistency in the results,
such as equal lengths for the aligned query and subject sequences
(which often isn't the case when gaps are introduced)
or if the parser doesn't accept the hyphens that are used to represent gaps.
In any case,
if gapped output simply can not be tolerated by one's parser,
one can still take advantage of the bug fixes,
improved speed, and other features of WU&nbsp;BLAST&nbsp;2.0
by using the <B>compat1.4</B> compatibility option.
</LI>

<LI>
Gapped alignments in the <B>blastn</B> search mode
are evaluated correctly, using different statistical parameter values
(<I>&lambda;</I>, <I>K</I> and <I>H</I>)
than those used
to evaluate the significance of ungapped alignments.
This is indeed the case for all WU&nbsp;BLAST search modes.
And if the correct parameters are not available for the particular
combination of gap penalties and scoring system being used,
a prominent warning to this effect is displayed.
</LI>

<LI>
Unique to WU&nbsp;<B>blastn</B> is support for fully-specified scoring matrices,
not just simple match/mismatch scoring systems.
This allows (for example) transitions to be scored differently than transversions;
and permits positive G-A substitution scores
for the design of siRNAs (small interfering RNAs)
where G-U base pairing is allowed.
Scoring matrices may also be tailored to improve the design of PCR primers.
Contrary to
<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=11331233&amp;dopt=Abstract" TARGET="_blank">W.&nbsp;Miller&nbsp;(2001)</A>,
scoring matrices were first supported in 1994, by the NCBI's ungapped <TT>BLASTN</TT> version 1.4
(Gish, W., unpublished;
see <A HREF="http://blast.wustl.edu/blast-1.4">http://blast.wustl.edu/blast-1.4</A>).
Support for nucleotide scoring matrices was dropped
by the NCBI's <B>blastall</B> 2.0 program first released in 1997,
but has been maintained continuously in all WU versions of the software
since the migration to Washington University in 1994.
</LI>

<LI>
Word lengths (re: the <I>W</I> parameter) as short as 1 have been
supported continuously by WU <B>blastn</B>,
as are nucleotide neighborhood words,
using the neighborhood word score threshold parameter, <I>T</I>.
Using neighborhood words, nucleotide sequence similarity can be detected
even in the absence of any identical residues between two sequences.
Users are cautioned, however, that careless use of the <I>T</I> parameter
can result in vast and overwhelming amounts
of memory being requested by the software;
<I>T</I> should likely be used only in conjunction
with very short word lengths.
</LI>

<LI>
Information describing &quot;consistent&quot; groups of alignments (HSPs) is provided
by licensed BLAST&nbsp;2.0, when the <B>topcomboN</B> or <B>links</B> options are used.
This facility can help with construction of distinct gene structure(s) from a barrage of alignments.
</LI>

<LI>
Licensed WU&nbsp;BLAST&nbsp;2.0 supports the eXtended Database Format (<B>XDF</B>),
a power user's dream in many ways
for working with peptide and nucleotide sequences.
Both the NCBI&nbsp;BLAST&nbsp;2.0 database format and the NCBI implementation
of the BLAST search algorithm are restricted to sequences under 16 Mbp
in length,
whereas human genome contigs exceeded 25 Mbp in the previous millennium
(<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=10830953&amp;dopt=Abstract" TARGET="_blank">Hattori&nbsp;<I>et&nbsp;al.</I>,&nbsp;2000</A>)
and extend to hundreds of megabytes today.
In contrast, <B>XDF</B> can accurately store individual sequences
of up to 1 Gbp (billion bp) with ambiguity codes intact.
Other BLAST software, such as the NCBI's,
limits database files to 2 gigabytes,
whereas WU&nbsp;BLAST's XDF supports databases (and database files) of virtually unlimited size &mdash;
provided of course that the underlying
operating system supports these so-called &ldquo;large files&rdquo;
(as most modern operating systems do).
</LI>
</UL>

<P>
In support of XDF,
a new database formatting tool named <B>xdformat</B>
is provided in the WU&nbsp;BLAST&nbsp;2.0 package.
Among other distinct capabilities
and advantages to using XDF and <B>xdformat</B> are:
</P>
<UL>
<LI>
fast appends of new sequences to existing databases
(both protein and nucleotide) &mdash;
no need to reformat
a database completely just to add more sequence(s) to it.
</LI>
<LI>
<B>xdformat</B> runs twice as fast as the NCBI's formatdb program,
while offering a vast superset of features and greater reliability;
</LI>
<LI>
safe roll-backs of database updates when file I/O
(<I>e.g.</I>, disk full errors) or parse errors are encountered;
</LI>
<LI>
huge databases need not be broken into multiple volumes
individually composed of several files but can be managed simply
with just 3 files, regardless of the database size;
</LI>
<LI>
flexible indexing of <I>all</I> sequence identifiers
&mdash; not just a subset of the NCBI identifiers &mdash;
including user-defined identifiers;
</LI>
<LI>
index support for duplicate occurrences of the same identifier,
even the identical "gi" identifiers that cause some indexing programs
to abort;
</LI>
<LI>
identifier indexing is supported not only when creating an XDF database
but when appending new sequences to an existing XDF database;
and if a database was originally created without an identifier index,
one can be added later in one fast step;
</LI>
<LI>
identifier indexes can be quickly re-built
if necessary using different indexing policies,
without having to reformat the entire database;
</LI>
<LI>
intelligent retrieval of indexed sequences
using a complementary program named <B>xdget</B>;
<B>xdget</B> can retrieve sequences by identifier
even if is not told what name space
(<I>e.g.</I>, gi, accession, locus, user-defined, etc.)
the identifier came from;
for more on identifier indexing, see
<a href="http://blast.wustl.edu/doc/FAQ-Indexing.html" TARGET="_blank">this</A>;
</LI>
<LI>
<B>xdformat</B> and <B>xdget</B> accept and work intelligently
with identifiers that obey the International DDBJ/EBI/NCBI collaboration's
Accession<B>.</B>Version identifier syntax (<I>e.g.</I>,
the programs know that BAA84643.2 is a newer version of BAA84643,
but will retrieve BAA84643.1 if specifically requested);
And the parsing programs that come with WU BLAST
for converting GenBank and EMBL database flat files into "FASTA" format
not only report <I>gi</I> identifiers but Accessions <B><I>with Versions</I></B>;
</LI>
<LI>
greatly reduced memory requirements and search initiation times
for databases containing large numbers of entries,
which is particularly important when memory is in short supply
or when multiple processors are standing by
waiting for a single-threaded initialization phase to be completed;
</LI>
<LI>
the ability to dump (or recover) the contents of an XDF database back into FASTA format
with the original annotation and ambiguity codes intact;
</LI>
<LI>
both the <B>X</B> and <B>N</B> ambiguity codes are supported in nucleotide sequences,
thus permitting the use of alternative substitution
scores for these letters and the use of PHRED/PHRAP sequence output "as is" as input
to <B>xdformat</B>.
</LI>

<LI>
Compared to the classical BLAST&nbsp;1.4 database format,
XDF provides
the ability to use FASTA/Pearson format
input files with unjustified (<I>i.e.,</I> ragged or blank) input lines.
With nucleotide sequence databases,
there is also no longer the need to retain the original FASTA input file
in order to access the ambiguity codes during a database search.
</LI>

<LI>
Support for XDF by the BLAST search programs
does not come at the expense of backward compatibility.
WU BLAST can search databases in either XDF or the classical BLAST&nbsp;1.4
database formats.
Furthermore, by simply installing new versions of "setdb" and "pressdb",
the migration to using XDF can be performed swiftly and transparently,
without making any changes whatsoever to existing database maintenance scripts.
While providing this drop-in upgrade path to XDF,
support for legacy databases in the BLAST&nbsp;1.4 database format
is retained transparently, as well:
the WU BLAST search programs automatically
identify the database format being used
and adjust their operations accordingly.
This allows users to migrate incrementally to XDF,
at their own pace and as they see fit,
without losing the ability to study or reproduce results
obtained with older databases.
Even so, users are encouraged to make the migration to XDF,
as there are definite benefits to the new format,
including an improved nucleotide sequence data representation
and the ability to index sequences by their identifiers.
More information about the simplicity of migrating to XDF is provided
to licensed users of WU&nbsp;BLAST&nbsp;2.0.
</LI>

<LI>
When searching very large databases,
virtual memory requirements are dramatically reduced in WU&nbsp;BLAST&nbsp;2.0,
eliminating program failures that occurred
when system resource limits were unexpectedly reached.
</LI>

<LI>
Virtual databases are supported in licensed BLAST&nbsp;2.0.
Virtual databases can be
specified on the command line as a white space-delimited list of
component database names.
Virtual databases can be comprised of
components in either XDF <I>or</I> classical BLAST&nbsp;1.4 format,
although both formats are not supported at the same time.
For example:
<PRE>

  blastn "pri rod mam vrt htg" myquery.nt

</PRE>
</LI>

<LI>
With licensed BLAST&nbsp;2.0,
virtually no file size limits exist for databases and other files,
<I>unless</I> the underlying operating system does not support large files.
Operating systems such as Linux (kernel version 2.2 and earlier)
for 32-bit Intel computing platforms
are often incapable of using files larger than 2 GB,
although virtual database support (see above) helps avoid this limitation
by allowing large databases to be segmented into files of a manageable size.
Linux users in need of large file support should use a version 2.4.* kernel.
</LI>

<LI>
Licensed BLAST&nbsp;2.0 supports <I>segmented query sequences</I>,
such as the contigs that result from shotgun sequencing assembly
or perhaps multiple short probes for a given gene.
For example,
all of the contigs from a given clone can be concatenated together with a single hyphen (-) character
to delimit each contig.
Segment boundaries are therefore clearly distinguishable from purely ambiguous regions
of the sequence, while consuming little storage.
BLAST&nbsp;2.0 honors segment boundaries by guaranteeing that no alignment, be it ungapped or gapped,
will cross a boundary.
Support for segmented database sequences is in progress.
</LI>

<LI>
Multi-sequence query files are supported in licensed BLAST&nbsp;2.0,
such that every sequence in the FASTA file is searched against the specified database.
Previous versions of
the software only compared the first sequence in the query file against the database.
Each search result is separated from the next by a single ASCII form feed
character (control-L).  See the new <B>qrecmin</B> and <B>qrecmax</B> options.
</LI>

<LI>
The format of all dates reported in BLAST output can be controlled by the
UN*X standard <CODE>CFTIME</CODE> environment variable.
For example, dates will be reported in
<A HREF="http://www.google.com/search?q=iso+8601" TARGET="_blank">ISO&nbsp;8601</A>
standard format,
if <CODE>CFTIME</CODE> is set to <CODE>'%Y-%m-%dT%H:%M:%S'</CODE>.
Date strings produced by the
<TT>xdformat</TT> program are also governed by <CODE>CFTIME</CODE>.
Note that the format of many date and time strings reported for
XDF databases is determined by the setting (if any)
of <CODE>CFTIME</CODE> <I>when the database was created or last modified</I>.
</LI>

<LI>
Both <I>sequence filtering</I> and <I>word masking</I> of query sequences
are supported.
The terms "filter" and "mask" are sometimes used alone and interchangeably,
however there are two distinct techniques people can use
which deserve separate names.
Lower case alphabetic letters in the query sequence can be used to
inform the BLAST search program as to which residues it should either <I>filter</I>
(convert to X or N)
or <I>mask</I> (skip when generating neighborhood words but otherwise leave intact).
See the <B>lcfilter</B> and <B>lcmask</B> options, respectively.
</LI>

<LI>
Multiple <CODE>filter=&lt;filter&gt;</CODE> specifications can be requested
on the BLAST command line.
Each of the filters is executed independently
and their results are OR-ed at the end.
</LI>

<LI>
Whereas NCBI&nbsp;BLAST&nbsp;2.0 uses the original external filtering technique of BLAST&nbsp;1.3
(Gish, W., unpublished),
which utilizes the UNIX popen() system call and temporary files,
WU&nbsp;BLAST&nbsp;2.0 avoids these problematic system interfaces.
</LI>

<LI>
One or more <B>word masks</B> can be specified on the command line, using
the <CODE>wordmask=&lt;<I>mask</I>&gt;</CODE> option, where &lt;<I>mask</I>&gt; may be a classical
filter program such as <CODE>seg</CODE>, <CODE>xnu</CODE>, or <CODE>dust</CODE>.
Whereas sequence filters
convert certain letters in the query sequence into ambiguity codes (X
for amino acid and N for nucleotide), word masks do not alter the
sequence.  Word masks instead cause the indicated portion(s) of the query
sequence to be skipped during BLAST neighborhood word generation.
This leaves the query sequence intact for generating alignments that are
seeded by word hits arising in flanking, unmasked regions of the
sequence.
</LI>

<LI>
The BLAST algorithm's word length parameter,
<I>W</I>, can be set
from 1 to 1024 in <I>all</I> search modes
(<TT>BLASTP</TT>, <TT>BLASTN</TT>, <TT>BLASTX</TT>, <TT>TBLASTN</TT>, <TT>TBLASTX</TT>).
</LI>

<LI>
WU&nbsp;BLAST&nbsp;2.0 reliably supports parallel processing on a variety of SMP
(symmetric multiprocessing) computing platforms.
WU&nbsp;BLAST is the only BLAST that threads properly across multiple CPUs
on dual-processor Apple PowerMacs running Mac&nbsp;OS&nbsp;X
and does not require a G4 processor.
POSIX threads are used under
Compaq Tru64 UNIX&nbsp;4.0+, Linux for X86 and Alpha processors,
IBM&nbsp;AIX, Mac&nbsp;OS&nbsp;X, IRIX&nbsp;6.5, and HP/UX&nbsp;11.
While POSIX threads are available under Solaris&nbsp;2+
(SPARC and X86),
Solaris threads are specifically used instead for slightly better performance.
The IRIX m_fork() system call provides parallel processing
under older versions of IRIX&nbsp;4 through 6.4;
and DCE threads are used under Digital UNIX&nbsp;3.2.
</LI>

<LI>
To illustrate just some of the flexibility available in WU&nbsp;BLAST&nbsp;2.0,
the licensed package includes a
<A HREF="http://www.cpan.org">PERL</A> script named <B>wu-blastall</B>
that translates an NCBI blastall command line
into a <I>rough</I> equivalent WU&nbsp;BLAST
command line and then invokes the appropriate WU&nbsp;BLAST search mode.
The output remains in WU&nbsp;BLAST format, but the <B>wu-blastall</B> script may help
users of the NCBI blastall program migrate to WU&nbsp;BLAST and start to discover its power.
<BR>
NOTE:
the <B>wu-blastall</B> script is <I>not</I> at all intended to provide
a literal replacement for the NCBI blastall program
and is <I>not</I> an appropriate method
for assessing the relative performance
(sensitivity, specificity, accuracy or speed)
of the NCBI and WU packages.
</LI>

<LI>
The <A HREF="http://sapiens.wustl.edu/maskeraid">MaskerAid</A>
substitute for CrossMatch (Phil Green,
<A HREF="http://www.genome.washington.edu/uwgc/analysistools/swat.htm">unpublished</A>)
provides another example of how the unique combination of flexibility and speed of WU&nbsp;BLAST&nbsp;2.0 can be applied
to yield 30-fold faster performance of
<A HREF="http://ftp.genome.washington.edu/RM/RepeatMasker.html">RepeatMasker</A> in its slow mode,
while maintaining sensitivity.
<B>MaskerAid</B> is licensed separately from WU&nbsp;BLAST.
</LI>

<LI>
Many additional command line options are available, some of which are described
<A HREF="#Options">below</A>.
Others are described in the README.html file bundled with the licensed software.
For users' convenience,
the licensed package includes compiled versions
of the
<A HREF="http://blast.wustl.edu/pub/nrdb" TARGET="_blank">
nrdb</A>
program, as well.
</LI>

</UL>

<P>
A reverse chronological list of changes is available in the <A HREF="http://blast.wustl.edu/blast/HISTORY">HISTORY</A>
file,
however
this file is outdated with respect to licensed versions of WU&nbsp;BLAST&nbsp;2.0.
The reader might also get the unfortunate
impression that WU&nbsp;BLAST&nbsp;2.0 is unreliable,
when in fact the licensed version has shown itself to be very robust.
Furthermore, any bugs that have been found have typically been fixed within 24 hours
of their being reported.
For current HISTORY information,
licensed users should consult the HISTORY file
that is bundled in licensed software distributions.
<BR>
<BR>Please send bug reports, questions, or suggestions to
<script language="JavaScript" type="text/javascript">
<!--
    var first = 'ma';
    var second = 'il';
    var third = 'to:';
    var address = '&#87;&#97;&#114;&#114;&#101;&#110;&#32;&#71;&#105;&#115;&#104;&#32;&#60;&#103;&#105;&#115;&#104;';
    var domain = '&#119;&#97;&#116;&#115;&#111;&#110;&#46;&#119;&#117;&#115;&#116;&#108;&#46;&#101;&#100;&#117;&#62;&#63;&#115;&#117;&#98;&#106;&#101;&#99;&#116;&#61;&#66;&#76;&#65;&#83;&#84;';
    document.write('<a href="');
    document.write(first+second+third);
    document.write(address);
    document.write('&#64;');
    document.write(domain);
    document.write('" title="Click here to compose an e-mail">');
    document.write('BLAST E-mail<\/a>');
// -->
</script>
<BR>
</P>

<HR>
<A NAME="Manifest"></A>
<H3>Manifest</H3>

<P>
The licensed BLAST&nbsp;2.0 package
includes the following data analysis and utility programs:
</P>

<UL>
<LI>
<B>blasta</B> - the unified database search program, which provides
<B>blastp</B>, <B>blastn</B>, <B>blastx</B>, <B>tblastn</B>, and <B>tblastx</B>
search functionality.
</LI>

<LI>
<B>xdformat</B> - the recommended program for rapidly converting sequences from
FASTA format into the native <B>XDF</B> format read by <B>blasta</B>.  The program
can also append new sequences to an existing database;
automatically rollback on errors;
provides flexible indexing and verification services;
and can dump data back into FASTA format.
</LI>

<LI>
<B>xdget</B> - a flexible tool for retrieving sequences
(or segments thereof)
from an indexed XDF database;
retrieved sequences are
optionally reverse-complemented and translated
in the case of nucleotide sequences.
xdformat and xdget are actually one-and-the-same program,
to ensure their compatibility.
</LI>

<LI>
<B>nrdb</B> - a tool for rapidly removing trivial redundancy (<I>i.e.</I>, duplicate
sequences) from one or more input files in FASTA format.  A simple hash table
is used, combined with data compression techniques to allow larger nucleotide
sequence data sets to be manipulated in memory.
</LI>

<LI>
<B>patdb</B> - a tool for rapidly removing trivial redundancy,
as well as identifying <I>perfect substrings</I>,
from one or more input files
in FASTA format.
A Patricia Tree is used, combined with a Finite State Automaton.
This tool is perhaps most useful when applied to protein sequences,
which often differ in their inclusion of the initiator methionine
or other post-translational modifications.
<B>Patdb</B> can also be more practically applied to protein
sequences than to nucleotide sequences,
because the data compression techniques of the <B>nrdb</B> program,
which are so effective with nucleotide sequences,
are not employed by <B>patdb</B>.
</LI>

<LI>
<B>wu-blastall</B> - a PERL script for converting an NCBI <B>blastall</B> command line
into a <I>rough</I> equivalent <B>blasta</B> command line and then invoking <B>blasta</B>.
The output is still in WU BLAST format.
This is primarily intended as a technology demonstration tool
but may also assist users in their migration
from NCBI BLAST to the more accurate WU BLAST.
For benchmarking of BLASTs,
careful tweaking of parameters may be required, but even with great care,
benchmarking for speed can still be confounded by inaccuracies in NCBI BLAST.
</LI>

<LI>
<B>wu-formatdb</B> - a PERL script for converting an NCBI <B>formatdb</B> command line
into the equivalent <B>xdformat</B> command line and then invoking <B>xdformat</B>.
This is primarily intended as a technology demonstration tool but may also
assist users in their migration from NCBI BLAST to WU BLAST.
</LI>

<LI>
<B>pam</B> - a program to compute amino acid substitution scoring matrices
having arbitrary scales, using the Dayhoff PAM model.
</LI>

<LI>
<B>pressdb.real</B> - the legacy <B>pressdb</B> program for users who are reliant
on the NCBI&nbsp;BLAST&nbsp;1.4 database format for nucleotide sequences.
</LI>

<LI>
<B>setdb.real</B> - the legacy <B>setdb</B> program for users who are reliant
on the NCBI&nbsp;BLAST&nbsp;1.4 database format for amino acid sequences.
</LI>

<LI>
<B>gb2fasta</B> - a parser to extract nucleotide sequences from GenBank flat files
into FASTA format.
</LI>

<LI>
<B>gt2fasta</B> - a parser to extract amino acid sequences from CDS features
in GenBank flat files and output them in FASTA format.
</LI>

<LI>
<B>sp2fasta</B> - a parser to extract protein or nucleotide sequences from
EMBL, TrEMBL, or SWISS-PROT database files and output them in FASTA format.
</LI>

<LI>
<B>pir2fasta</B> - a parser to extract protein sequences from NBRF PIR database
files and output them in FASTA format.
</LI>

<LI>
<B>seg</B> - a low-complexity filter for protein and nucleotide sequences
(<A HREF="#Wootton_and_Federhen_1993">Wootton and Federhen, 1993</A>;
<A HREF="#Wootton_and_Federhen_1996">Wootton and Federhen, 1996</A>).
The program identifies low compositional complexity regions.
</LI>

<LI>
<B>dust</B> - a low-complexity filter for nucleotide sequences
(<A HREF="#Hancock_and_Armstrong_1994">Hancock and Armstrong, 1994</A>;
Tatusov and Lipman, unpublished).
</LI>

<LI>
<B>xnu</B> - a low-complexity filter for protein sequences
(<A HREF="#Claverie_and_States_1993">Claverie and States, 1993</A>).
The program identifies short-periodicity repeats.
</LI>


<LI>
<B>sysblast.sample</B> - a sample configuration file that system
administrators may wish to modify and install as <B>/etc/sysblast</B>.
Parameter settings in this file can be used to:
limit the number of threads employed by each BLAST process;
change the default number of threads employed per process;
alter the "nice" value for BLAST processes;
limit the amount of memory utilized by each BLAST process.
</LI>

</UL>

<HR>

<A NAME="Tofly"></A>
<H3>To Fly...</H3>

<P>
If the gapped alignments are nice, but even more speed or less memory use are desired,
read how to make the programs <A HREF="http://blast.wustl.edu/blast/TO-FLY.html">fly</A>.
<BR>
</P>

<HR>
<A NAME="Examples"></A>
<H3>Examples</H3>

<P>
Here are some sample WU&nbsp;BLAST&nbsp;2.0 results
produced using generally default parameters,
plus the oft-recommended low-complexity filter <B>seg</B>
and the <B>-postsw</B> option of WU BLASTP 2.0.
Exceptions to the defaults are noted and their corresponding results provided,
as well.
<BR>
Default parameters for NCBI blastall were also used,
with the exception of using -G7 -E2
to make the scoring system identical to the WU default
(penalty of 9 for the first residue in a gap).
</P>

<UL>
<LI>
<B>Example&nbsp;1</B> produced
<A HREF="http://blast.wustl.edu/blast/example1-20postsw.html" TARGET="_blank">
with</A>
and
<A HREF="http://blast.wustl.edu/blast/example1-14.html" TARGET="_blank">
without</A>
gaps;
<I>with</I> gaps but
<A HREF="http://blast.wustl.edu/blast/example1-20kap.html" TARGET="_blank">
without</A>
Karlin-Altschul "Sum" statistics;
and <I>with</I> gaps and Karlin-Altschul "Sum" statistics but
<A HREF="http://blast.wustl.edu/blast/example1-20.html" TARGET="_blank">
without</A>
the <B>-postsw</B> option.
<BR>
Example 1 NCBI blastall
<A HREF="http://blast.wustl.edu/blast/example1-ncbi.html" TARGET="_blank">output</A>.
</LI>

<LI>
<B>Example&nbsp;2</B> produced
<A HREF="http://blast.wustl.edu/blast/example2-20.html" TARGET="_blank">
with</A>
and
<A HREF="http://blast.wustl.edu/blast/example2-14.html" TARGET="_blank">
without</A>
gaps;
and <I>with</I> gaps but
<A HREF="http://blast.wustl.edu/blast/example2-20kap.html" TARGET="_blank">
without</A>
Karlin-Altschul "Sum" statistics.
<BR>
Example 2 NCBI blastall
<A HREF="http://blast.wustl.edu/blast/example2-ncbi.html" TARGET="_blank">output</A>.
</LI>

<LI>
<B>Example&nbsp;3</B> produced
<A HREF="http://blast.wustl.edu/blast/example3-20.html" TARGET="_blank">
with</A>
and
<A HREF="http://blast.wustl.edu/blast/example3-14.html" TARGET="_blank">
without</A>
gaps.
<BR>
Example 3 NCBI blastall
<A HREF="http://blast.wustl.edu/blast/example3-ncbi.html" TARGET="_blank">output</A>.
</LI>

</UL>


<HR>
<A NAME="Options"></A>
<H3>New Command Line Options</H3>

<P>
Command line options for WU&nbsp;BLAST version 1.4
often apply to version 2.0 without change.
(See the version
<A HREF="http://blast.wustl.edu/blast-1.4/blast1.pdf">1.4 manual page</A>
in
<A HREF="http://www.adobe.com/prodindex/acrobat/">Adobe Acrobat</A> (PDF) format).
New command line options to version 2.0 include the following.
Some of these options are not available in the freely available <I>alpha</I> releases
of WU&nbsp;BLAST&nbsp;2.0.
Terse program usage can also be obtained by entering one
of the program names on the command line without any arguments.
<BR>
Note: parsing of command line options is alphabetic case-<I>in</I>dependent.
</P>

<TABLE BORDER="2" CELLPADDING="2">
	<TR>
		<TD WIDTH="89">Option</TD>
		<TD WIDTH="279">Description</TD>
	</TR>
	<TR>
		<TD>Q=&lt;<I>q</I>&gt;</TD>
		<TD>set the penalty for a gap of length one to <I>q</I> (default Q=9 for proteins; Q=10 for <TT>BLASTN</TT>)</TD>
	</TR>
	<TR>
		<TD>R=&lt;<I>r</I>&gt;</TD>
		<TD>set the per-residue penalty for extending a gap to <I>r</I> (default R=2 for proteins; R=10 for <TT>BLASTN</TT>)</TD>
	</TR>
	<TR>
		<TD>H=&lt;<I>h</I>&gt;</TD>
		<TD>Set the value for the relative entropy to be used in Karlin-Altschul statistics of <I>un</I>gapped alignment scores.
		In earlier versions of BLAST, the <I>H</I> option was used to invoke the display of a histogram.
		</TD>
	</TR>
	<TR>
		<TD>postsw</TD>
		<TD>perform full Smith-Waterman alignment of sequences and re-rank the database matches accordingly, prior to output (currently supported in <TT>BLASTP</TT> only)</TD>
	</TR>
	<TR>
		<TD>hitdist=&lt;<I>hitdist</I>&gt;</TD>
		<TD>invoke a 2-hit BLAST algorithm that is similar to but more sensitive than that of <A HREF="#Altschul_et_al_1997">Altschul <I>et al.</I> (1997)</A>, with the maximum distance between word hits of &lt;<I>hitdist</I>&gt;.
<A HREF="#Altschul_et_al_1997">Altschul <I>et al.</I> (1997)</A> use the equivalent of hitdist=40 in the <TT>BLASTP</TT>, <TT>BLASTX</TT>, <TT>TBLASTN</TT> and <TT>TBLASTX</TT> search modes.
 In WU&nbsp;BLASTN, setting hitdist=<I>W</I> and wink=<I>W</I>, where <I>W</I> is the word length, is akin to using double-length words generated
on <I>W</I>-mer boundaries.
		<BR>
		<I>NOTE:</I> in protein-level comparisons, for best sensitivity (or the best sensitivity for the amount of memory used), 2-hit BLAST should generally be avoided.
	<BR>
	This option is only available in the licensed 2.0 software.
	</TD>
	</TR>
	<TR>
		<TD>wink=&lt;<I>wink</I>&gt;</TD>
		<TD>generate word hits at every <I>wink</I>th (&quot;W increment&quot;) position along the query,
where the default wink=1 produces neighborhood words at every position.
For best sensitivity,
this option (setting wink greater than 1) should not be used.
Wink is best used to find identical or nearly identical sequences rapidly.
When used in conjunction with the hitdist option
to obtain the highest search speed, care should be taken that desirable alignments are not precluded
by these parameters.
	This option is only available in the licensed 2.0 software.
	</TD>
	</TR>
	<TR>
		<TD>wordmask=&lt;<I>masker</I>&gt;</TD>
		<TD>mask letters in the query sequence without altering the sequence itself,
		during neighborhood word generation.
		</TD>
	</TR>
	<TR>
		<TD>lcfilter</TD>
		<TD>filter lower case letters in the query sequence,
		by replacing lower case letters with the appropriate ambiguity code
		(N for nucleotide sequences, X for protein sequences).
		</TD>
	</TR>
	<TR>
		<TD>lcmask</TD>
		<TD>mask lower case letters in the query sequence <I>without altering the sequence</I>,
		during neighborhood word generation.
		</TD>
	</TR>
	<TR>
		<TD>maskextra=&lt;<I>extra</I>&gt;</TD>
		<TD>word-mask an additional <I>extra</I> letters on each side flanking an already-masked region.  This helps avoid the appearance
		 of spurious alignments through low-complexity regions initiated by chance word hits immediately adjacent to masked regions.
		 </TD>
	 </TR>
	<TR>
		<TD>nogaps</TD>
		<TD>do not create gapped alignments and, in essence, revert to WU&nbsp;BLAST&nbsp;1.4 behavior</TD>
	</TR>
	<TR>
		<TD>wstrict</TD>
		<TD>
		require that all ungapped alignments found during the initial, ungapped phase of a search
		contain an identical word hit (in the usual case of <TT>BLASTN</TT> usage)
		or a neighborhood word hit (in the case of <TT>TBLASTN</TT> and <TT>TBLASTX</TT>),
		when searching a nucleotide database sequence that contains one or more
		ambiguous residues.
		The <B>-wstrict</B> option has no effect whatsoever on <TT>BLASTX</TT>
		and has no effect on <TT>BLASTP</TT> when gapped alignments (the default)
		are to be produced.
		</TD>
	</TR>
	<TR>
		<TD>pingpong</TD>
		<TD>Perform additional work to help ensure the alignments produced are locally optimal.
		This option typically adds 3-10% to the execution time, without affecting
		the results.
		Only <I>rarely</I> is an alignment and its associated score improved, for all the work involved.
		</TD>
	</TR>
	<TR>
		<TD>gapall</TD>
		<TD>effectively generate a gapped alignment for every ungapped HSP found.  This is the default behavior.
		<BR>
		<I>See also:</I> <B>gapE</B>.</TD>
	</TR>
	<TR>
		<TD>gapE=&lt;<I>gapE</I>&gt;</TD>
		<TD>generate gapped alignments for all HSPs between sequences whose expected frequency of chance occurrence is less than or equal to &lt;<I>gapE</I>&gt;.&nbsp;Default value is gapE=<I>infinity</I>, <I>i.e.,</I> <B>gapall</B> is in effect.</TD>
	</TR>
	<TR>
		<TD>gapW=&lt;<I>gapW</I>&gt;</TD>
		<TD>set the window width (or band width) within which gapped alignments are generated (default is gapW=32 for protein comparisons, gapW=16 for <TT>BLASTN</TT>).</TD>
	</TR>
	<TR>
		<TD>noseqs</TD>
		<TD>produces greatly abbreviated output that omits sequence alignments and yet may be interpreted correctly by existing parsers.</TD>
	</TR>
	<TR>
		<TD>hspmax=&lt;<I>hspmax</I>&gt;</TD>
		<TD>establishes <I>&lt;hspmax&gt;</I> as the maximum number of ungapped HSPs
		that will be saved per subject sequence or pairwise sequence comparison.
		Saved HSPs are then fed to the gapped alignment phase of the program
		or are statistically evaluated
		if gapped alignments are not to be performed.
		If more than <I>&lt;hspmax&gt;</I> HSPs are found,
		only the best-scoring HSPs are retained for subsequent processing.
		<BR>
		The default value is 1000; a value of 0 implies no limit.
		<BR>
		<I>See also:</I> <B>gspmax</B> and <B>spoutmax</B>.
		<BR>
		<I>NOTE:</I> this usage of <B>hspmax</B> is subtly, but importantly,
		different from the parameter's classical interpretation,
		wherein all ungapped HSPs that satisfied the S2 score threshold
		were saved and <I>&lt;hspmax&gt;</I> merely limited
		the number of HSPs (gapped or ungapped) that would be reported.
		The new interpretation was instituted to provide vastly improved speed
		on large problems,
		while imparting no effect on small problems
		and many medium-sized problems.
		The new behavior can help guard against horrendously slow searches
		resulting from an inadvertant omission of a low-complexity filter.
		Adverse effects on sensitivity may be obtained, however,
		if every HSP is sacred.
		To restore classical behavior, specify <I>hspmax=0</I>.
		As a compromise between sensitivity and speed, set a higher
		value than the default.
		<BR>
		<I>NOTE:</I> the <B>B</B> and <B>V</B> options limit the number
		of subject sequences for which any results whatsoever are reported,
		regardless of the number of HSPs or GSPs found in each case.
		</TD>
	</TR>
	<TR>
		<TD>gspmax=&lt;<I>gspmax</I>&gt;</TD>
		<TD>establishes <I>&lt;gspmax&gt;</I> as the maximum number of GSPs (gapped HSPs)
		to report per subject sequence or pairwise sequence comparison.
		If more than <I>&lt;gspmax&gt;</I> GSPs are found,
		only the best-scoring GSPs are retained for subsequent processing and reporting.
		The setting of <B>gspmax</B> will have no effect,
		if the <B>nogaps</B> option is specified or
		if the setting of <B>hspmax</B> is more restrictive.
		<BR>
		The default value for <B>gspmax</B> is 0;
		a value of 0 implies no limit.
		<BR>
		<I>See also:</I> <B>hspmax</B> and <B>spoutmax</B>.
		<BR>
		<I>NOTE:</I> the <B>B</B> and <B>V</B> options limit the number
		of subject sequences for which any results whatsoever are reported,
		regardless of the number of HSPs or GSPs found.
		</TD>
	</TR>
	<TR>
		<TD>spoutmax=&lt;<I>spoutmax</I>&gt;</TD>
		<TD>establishes <I>&lt;spoutmax&gt;</I>
		as the maximum number of segment pairs to report
		in program output per subject sequence or pairwise comparison,
		independent of the number of HSPs or GSPs actually found and evaluated.
		If more than <I>&lt;spoutmax&gt;</I> segment pairs are found,
		the segment pairs are sorted by the sort criteria in effect
		for the search and only the first <I>&lt;spoutmax&gt;</I>
		segment pairs will be reported.
		The setting of <B>spoutmax</B> will have no effect
		if either <I>&lt;hspmax&gt;</I> or <I>&lt;gspmax&gt;</I>
		is more restrictive.
		<BR>
		The default value for <B>spoutmax</B> is 0;
		a value of 0 implies no limit.
		<BR>
		<I>See also:</I> <B>hspmax</B> and <B>gspmax</B>.
		</TD>
	</TR>
	<TR>
		<TD>compat1.4</TD>
		<TD>produces BLAST version 1.4-style output (no gaps), but with bug fixes and performance enhancements in place.</TD>
	</TR>
	<TR>
		<TD>kap</TD>
		<TD>use Karlin-Altschul (1990) statistics on individual alignment scores (<I>i.e.,</I> do not evaluate the joint probability of multiple scores, such as with Poisson or Karlin-Altschul (1993) "Sum" statistics).</TD>
	</TR>
	<TR>
		<TD>restest</TD>
		<TD>causes statistical significance estimates to depend upon the size of the database,
		as determined by the total number of <I>residues</I> it contains.
		<B>Restest</B> is the default method for determining the database size
		in the <B>blastn</B>, <B>tblastn</B>, and <B>tblastx</B> search modes.
		<BR>
		See <B>seqtest</B>.
		</TD>
	</TR>
	<TR>
		<TD>seqtest</TD>
		<TD>causes statistical significance estimates to depend upon the size of the database,
		as determined by the number of <I>sequences</I> it contains.
		<B>Seqtest</B> is the default method for determining the database size
		in the <B>blastp</B> and <B>blastx</B> search modes.
		<I>For backward compatibility with legacy BLAST software
		&mdash; in all search modes, including <B>blastp</B> and <B>blastx</B> &mdash;
		if the <B>Z</B> option is specified,
		<B>Z</B> is expected to be expressed in units of residues,
		unless the <B>seqtest</B> option is also specified.</I>
		<BR>
		See <B>restest</B>.
		</TD>
	</TR>
	<TR>
		<TD>links</TD>
		<TD>display consistent link information for each HSP,
		indicating which &ldquo;consistent&rdquo;
		alignments were used in joint
		statistical significance calculations
		(<I>i.e.,</I> Sum or Poisson statistics).
		NOTE: while all link lists describe mutually consistent HSPs,
		unless one of the &ldquo;topcombo&rdquo; options is used,
		only lists reported for HSPs in the most significant set
		for each database sequence are guaranteed to represent the
		precise set of HSPs for which the joint statistics were computed;
		all other link lists often correctly describe the set of HSPs
		involved but could have one or more missing or extraneous HSPs.
		 </TD>
	</TR>
	<TR>
		<TD>topcomboN=&lt;<I>n</I>&gt;</TD>
		<TD>report at most <I>n</I> &quot;topcombo&quot; groups of consistent (colinear) local alignments (HSPs).
		Each local alignment is allowed to be a member of only one group.
		Use of this option causes the addition of a "Group = #" indicator
		in the output for each HSP.
		Groups of HSPs tend to be assembled in decreasing order of statistical
		significance.
		Members of the most significant group thus tend to be reported
		with "Group = 1".
		<I>See also:</I> <I><B>topcomboE</B></I>.
		</TD>
	</TR>
	<TR>
		<TD>topcomboE=&lt;<I>E<sub>ratio</sub></I>&gt;</TD>
		<TD><I>E<sub>ratio</sub></I> is the maximum ratio of <I>E<sub>current</sub>/E<sub>best</sub></I> for which
		the <I>current</I> &quot;topcombo&quot; group of consistent (colinear) local alignments will be reported
		for a given database sequence.
		The "best" group is reported in the output as "Group = 1"
		and tends to be the most statistically significant.
		The default behavior is to impose no limit on this ratio, in which case all topcombo groups satisfying E are reported (up to a maximum of <I><B>topcomboN</B></I>).
		<I>See also:</I> <B>topcomboN</B>.
		</TD>
	</TR>
	<TR>
		<TD>olmax=&lt;<I>len</I>&gt;</TD>
		<TD>maximum permitted length of overlap (in residues), <I>len</I>, of two <I>un</I>gapped alignments for their joint (Sum or Poisson) probability to be computed.  The default is unlimited length, with the maximum extent of overlap being governed only by the <I><B>olf</B>raction</I> parameter.</TD>
	</TR>
	<TR>
		<TD>golmax=&lt;<I>len</I>&gt;</TD>
		<TD>maximum permitted length of overlap (in residues), <I>len</I>, of two <I>gapped</I> alignments for their joint (Sum or Poisson) probability to be computed.  The default is unlimited length, with the maximum extent of overlap being governed only by the <I><B>golf</B>raction</I> parameter.</TD>
	</TR>
	<TR>
		<TD>hspsepQmax=&lt;<I>d</I>&gt;</TD>
		<TD>maximum allowed separation along the query sequence between two HSPs (gapped or ungapped)
		that will be clustered into a &quot;consistent&quot; set.
		Useful for improving the statistical power of discrimination of clusters
		that have potential biological interest from random background,
		when the query is longer than the clustered features of interest.
		Depending on the specific search to be performed, a reasonable setting
		for this parameter when used with a contig sequence
		might be the expected maximum length of an intron in the species of interest.
		The units for hspsepQmax are measured in residues of the
		sequence type actually compared,
		<I>e.g.,</I> amino acid residues for a <TT>BLASTX</TT> query.
		</TD>
	</TR>
	<TR>
		<TD>hspsepSmax=&lt;<I>d</I>&gt;</TD>
		<TD>maximum allowed separation along the subject (database) sequence for two HSPs (gapped or ungapped) to be clustered into a consistent set.
		Useful for improving the statistical power of discrimination of clusters
		that have potential biological interest from random background,
		when the database contains sequences longer than the features of interest.
		Depending on the specific search to be performed, a reasonable setting
		for this parameter when used with a genomic sequence
		might be the expected maximum length of an intron in the species of interest.
		The units for hspsepSmax are measured in residues of the
		sequence type actually compared,
		<I>e.g.,</I> nucleotides for a <TT>BLASTN</TT> search and
		amino acid residues for a <TT>TBLASTN</TT> search.
		</TD>
	</TR>
	<TR>
		<TD>gapK=&lt;<I>k</I>&gt;</TD>
		<TD>set the value of the Karlin-Altschul statistics' <I>K</I> parameter to use when evaluating the significance of gapped alignment scores.  Useful when precomputed values are unavailable in the internal tables for the chosen scoring matrix and gap penalty combination.</TD>
	</TR>
	<TR>
		<TD>gapL=&lt;<I>l</I>&gt;</TD>
		<TD>set the value of the Karlin-Altschul statistics' <I>&lambda;</I> parameter to use when evaluating the significance of gapped alignment scores</TD>
	</TR>
	<TR>
		<TD>gapH=&lt;<I>h</I>&gt;</TD>
		<TD>set the value of the Karlin-Altschul statistics' <I>H</I> parameter to use when evaluating the significance of gapped alignment scores</TD>
	</TR>
	<TR>
		<TD>Qoffset=&lt;<I>i</I>&gt;</TD>
		<TD>adjust all query sequence coordinates in the output by the fixed
		quantity <I>i</I> (default <I>i</I>=0).
		</TD>
	</TR>
	<TR>
		<TD>Soffset=&lt;<I>i</I>&gt;</TD>
		<TD>adjust all subject sequence coordinates in the output by the fixed
		quantity <I>i</I> (default <I>i</I>=0).
		</TD>
	</TR>
	<TR>
		<TD>dbchunks=&lt;<I>nchunks</I>&gt;</TD>
		<TD>
		establishes the granularity of the database, as it is divided into
		slices for assignment to individual threads,
		to make more efficient use of all CPUs when multiple CPUs
		are employed for a given search.
		Higher values are appropriate when the database contains relatively
		few sequences and/or when the sequences vary greatly in length,
		composition or content (<I>e.g.,</I> genomic contigs).
		Lower values are appropriate when the database contains many
		sequences of comparable length
		(<I>e.g.,</I> the EST division of GenBank).
		The minimum assignable value is the number of threads employed,
		but this setting is ill-advised;
		the optimal value for any given search type is likely to be
		a large multiple of the number of threads employed
		(although it need not be an exact multiple).
		When searching mammalian genomic contigs, a good value may be 1000.
		The default value is 500.
		</TD>
	</TR>
	<TR>
		<TD>dbslice=<I>m</I>/<I>n</I><BR>
		dbslice=<I>a-b</I>/<I>n</I><BR>
		</TD>
		<TD>at run time, logically divide the database into <I>n</I> equivalent-sized slices and search only the <I>m</I>-th slice, where 1&nbsp;&lt;=<I>m</I>&nbsp;&lt;=<I>n</I>&nbsp;&lt;=&nbsp;100000.
		Alternatively, search slices <I>a</I> through <I>b</I> (inclusive),
		where 1&lt;=<I>a</I>&lt;=<I>b</I>&lt;=<I>n</I>.
		Slice size is determined merely by the number of sequence records contained within, not the sum total sequence length.
		In distributed computing environments,
		when the same database is to be searched repeatedly and
		database slices are to be assigned to individual compute nodes,
		overall throughput will likely benefit from consistently assigning
		the same slice(s) to the same nodes for each search;
		the improved efficiency results from the file caching activity
		typically performed by an operating system.</TD>
	</TR>
	<TR>
		<TD>qrecmin=&lt;m&gt;</TD>
		<TD>in a multi-sequence query file, start database searches using the query sequence numbered <I>m</I>.  (The first record is numbered <I>1</I>).</TD>
	</TR>
	<TR>
		<TD>qrecmax=&lt;n&gt;</TD>
		<TD>in a multi-sequence query file, end database searches with the query sequence numbered <I>n</I>.</TD>
	</TR>
	<TR>
		<TD>putenv="<I>NAME=VALUE</I>"</TD>
		<TD>
		in the local environment to the BLAST search program, set the environment variable named <I>NAME</I> to the value <I>VALUE</I>.</TD>
	</TR>
	<TR>
		<TD>endputenv</TD>
		<TD>for security in WWW server installations, where the command line may sometimes be left open to users, ignore any subsequent <B>putenv</B> options found on the command line during left-to-right parsing.</TD>
	</TR>
	<TR>
		<TD>getenv="<I>NAME</I>"</TD>
		<TD>
		display the value of the environment variable named <I>NAME</I>.  This may be useful for verifying that the settings of environment variables on a web server or in an analysis pipeline have been propagated all the way to the BLAST search program.</TD>
	</TR>
	<TR>
		<TD>endgetenv</TD>
		<TD>ignore any subsequent <B>getenv</B> options found on the command line during left-to-right parsing.</TD>
	</TR>
	<TR>
	<TD>cdb</TD>
		<TD>search nucleotide sequence databases in their <I>uncompressed</I> form.
		This option is only effective in the <TT>BLASTN</TT> search mode for word lengths > 6.
		See <B>ucdb</B>.
		</TD>
	</TR>
	<TR>
	<TD>ucdb</TD>
	<TD>search nucleotide sequence databases in their <I>un</I>compressed form,
	with any-and-all ambiguity codes in place.
	This option may be used to increase sensitivity
	in the presence of ambiguity codes,
	at the expense of memory and possibly speed.
	This is the standard behavior for word lengths &lt; 7,
	and is not recommended for use with the default or longer word lengths,
	particularly for longer sequences, due to the increased memory requirements;
	when comparing long sequences, however, if sufficient memory is available, use of this option can yield a significant increase in speed.
	This option offers improved sensitivity when searching databases in XDF format that contain ambiguity codes.
	The option is accepted by the software but offers no improvement in sensitivity for databases in the earlier BLAST 1.4 database format.
	(<TT>BLASTN</TT> search mode only).
	</TD>
	</TR>
	<TR>
		<TD>mmio</TD>
		<TD>turn off the use of memory-mapped I/O when reading database files.&nbsp;Use of this option will usually retard the search, particularly when multiple processors are being used, but it serves both to demonstrate the effectiveness of this form of I/O and to validate the I/O routines.  Note that no special daemon or support programs (such as the old <I>memfile</I> program) are required to take full advantage of memory-mapped I/O.</TD>
	</TR>
</TABLE>


<HR>
<A NAME="EnvVars"></A>
<H3>Environment Variables</H3>

<P>
As described below and elsewhere,
WU&nbsp;BLAST&nbsp;2.0 supports several environment variables
to adapt its behavior to different computing environments:
<B>BLASTDB</B>, <B>BLASTFILTER</B> and <B>BLASTMAT</B>.
To support dual WU/NCBI BLAST installations,
WU&nbsp;BLAST also supports the environment variables
<B>WUBLASTDB</B>, <B>WUBLASTFILTER</B> and <B>WUBLASTMAT</B>,
with the WU versions of these variables taking precedence
over the corresponding non-WU versions when both are set.
</P>

<P>
In WU&nbsp;BLAST&nbsp;2.0, the <B>BLASTDB</B> (or WUBLASTDB) environment variable
can be a list of one <I>or more</I> directory names in which the programs
are to look for database files.
In UNIX parlance, such an environment variable might be called a <I>path</I>
for the database files.
Directory names should be delimited from one another by a colon
(":") and listed in the order that they should be searched.
If the BLASTDB environment variable is not set, the programs use a default
path of <CODE>.:/usr/ncbi/blast/db</CODE>, such that the programs first look in the
current working directory (".") for the requested database
and then look in the <CODE>/usr/ncbi/blast/db</CODE> directory.
For backward compatibility with
programs that expect BLASTDB to be a single directory specification and
not a path, if the user has set a value for BLASTDB but omitted the current
working directory,
the version 2 programs will still look for database files
in the current working directory as a last resort.
</P>

<P>
The <B>BLASTFILTER</B> (or WUBLASTFILTER) environment variable
can be set to the directory containing the filter programs,
such as
<A HREF="http://blast.wustl.edu/pub/seg" TARGET="_blank">seg</A> and
<A HREF="http://blast.wustl.edu/pub/xnu" TARGET="_blank">xnu.</A>
The default directory for the filter programs is <CODE>/usr/ncbi/blast/filter</CODE>.
This usage is unchanged from version 1.4.
</P>

<P>
The <B>BLASTMAT</B> (or WUBLASTMAT) environment variable can be set to the parent
directory for all scoring matrix files.
The default directory for these files is <CODE>/usr/ncbi/blast/matrix</CODE>,
beneath which are nt and aa subdirectories for storing scoring matrix
files appropriate for nucleotide and amino acid alphabets.
This usage is unchanged from version 1.4.
</P>

<P>
For more information about environment variables, see the
<A HREF="#Installation">Installation</A> instructions.
</P>

<HR>

<A NAME="Filters"></A>
<H3>Filters and Masks</H3>
<P>
WU&nbsp;BLAST provides an highly flexible means
of applying both "hard" and "soft" masks to a query sequence,
supporting alternative, user-defined filter programs,
as well as non-standard parameters to the standard filters.
The <B>filter</B> (for hard masking) and
<B>wordmask</B> (for soft masking)
command line options provide the basic interface.
Multiple specifications of each type are acceptable
on the BLAST command line;
and individual filter and wordmask specifications may
consist of entire pipelines of commands.
</P>

<P>
For example, three filters are used in succession by this pipeline:
</P>

<PRE>
      filter="myfilter1 | myfilter2 | myfilter3 -x5 -"
</PRE>

<P>
The first two filters in this case are expecting to read their input from UN*X
standard input (also known as <I>stdin</I>),
whereas <I>myfilter3</I> apparently needs to be told (with the usual "-" or
hyphen argument)
to read data from stdin.
The standard output (<I>stdout</I>)
from <I>myfilter1</I> will be read via stdin
by <I>myfilter2</I>,
which in turn processes the query
before handing its results to <I>myfilter3</I>;
finally, <I>myfilter3</I>
reports its results to stdout,
which the BLAST program itself reads to obtain the fully masked sequence.
The final output from the filter pipeline is expected by the BLAST
program to be in FASTA format.
</P>

<P>
Instead of running all 3 filters in the above example as part of one
pipeline, they could instead be specified as separate filter options
like this:
</P>

<PRE>
    filter=myfilter1  filter=myfilter2  filter="myfilter3 -x5 -"
</PRE>

<P>
The same choice of running as a pipeline or running separately is available
for wordmasks, too.
And of course the two approaches can be combined on the same command line.
An advantage to using the pipeline approach is that all 3 filters
in the example above may complete a little bit faster,
because much of the I/O is avoided.
Furthermore,
when used in the pipeline,
there's no requirement that the output from <I>myfilter1</I>
and <I>myfilter2</I> actually be in FASTA format.
Those two programs could potentially pass any information between
themselves and to <I>myfilter3</I>.
The only absolute requirement is that <I>myfilter1</I> must read FASTA data
from stdin and <I>myfilter3</I> must output FASTA data (of the same length
as the query!) to stdout.
</P>

<P>
It should be noted that with some filter programs,
passing the query sequence sequentially through
a pipeline of filters may yield
a different result than processing the query independently with each filter
and OR-ing the results.
The script <B>seg+xnu</B> included in the <I>filter/</I> directory provides
an example with which to test this.
Specifying <I>filter=seg+xnu</I> on the BLAST command line
invokes a seg and xnu pipeline that is built-in to the search programs;
whereas specifying <I>filter="seg+xnu -"</I>
causes the <B>seg+xnu</B> script to be invoked on the query, which independently
executes <B>seg</B> and <B>xnu</B>, then ORs the separate results with <B>pmerge</B>.
(The <B>echofilter</B> option can be used to see the results of filtering displayed
in search program output).
While the built-in seg+xnu pipeline is historically the way these two filters have
been implemented,
the latter interpretation, as illustrated by the seg+xnu script with pmerge,
may be more desirable.
</P>

<HR>

<A NAME="Bugs"></A>
<H3>Bugs</H3>
<P>
The following list describes bugs that are known to exist in the WU&nbsp;BLAST&nbsp;2.0<B>a19</B>
binaries posted here.
These are all fixed in the licensed version 2.0 of WU&nbsp;BLAST,
in which there are no <I>known</I> bugs.
If you are a user of the licensed version and believe you see a bug,
please send a
<script language="JavaScript" type="text/javascript">
<!--
	var first = 'ma';
	var second = 'il';
	var third = 'to:';
	var address = '&#87;&#97;&#114;&#114;&#101;&#110;&#32;&#71;&#105;&#115;&#104;&#32;&#60;&#103;&#105;&#115;&#104;';
	var domain = '&#119;&#97;&#116;&#115;&#111;&#110;&#46;&#119;&#117;&#115;&#116;&#108;&#46;&#101;&#100;&#117;&#62;&#63;&#115;&#117;&#98;&#106;&#101;&#99;&#116;&#61;&#66;&#76;&#65;&#83;&#84;&#32;&#66;&#117;&#103;&#32;&#82;&#101;&#112;&#111;&#114;&#116;';
	document.write('<a href="');
	document.write(first+second+third);
	document.write(address);
	document.write('&#64;');
	document.write(domain);
	document.write('" title="Click here to compose an e-mail bug report">');
	document.write('BLAST Bug Report<\/a>');
// -->
</script>

Even users of the licensed version should read the second set
of potential problem areas or pitfalls listed below this initial bug list.
</P>

<UL>
<LI>
Due to a 2 GB file size limit associated with many 32-bit computing platforms
(<I>e.g.,</I> Solaris 2.5 and earlier, IRIX 5 and earlier, and many versions of Linux on X86),
users will be unable to search nucleotide sequence databases
where the accompanying FASTA file is larger than 2 GB.
"Loss-less" compression programs like
<A HREF="http://blast.wustl.edu/pub/nrdb" TARGET="_blank">nrdb</A>
can help keep the data size below the 2 GB limit
(while speeding up searches proportionately),
but this is merely a stop-gap solution,
as the public and private data sets are growing at exponential rates.
A comprehensive set of GenBank nucleotide sequences
in FASTA format currently (01-Oct-2001) exceeds 16 GB.
Switching to a 64-bit computing platform &mdash; or one which provides
"largefile" support, such as Solaris 2.6+ under SPARC or Intel processors &mdash;
would be sufficient to break the 2 GB barrier for these BLAST searches,
but additional problems in 2.0a19 still preclude its use beyond 2 GB.
</LI>
<LI>
The low-complexity sequence <B>-filter</B> option
causes fatal errors in the 2.0a19 alpha version of <B>blastn</B>.
</LI>
<LI>Use of the <B>-consistency</B> or <B>-poissonp</B> options (or both)
results in severely truncated output, as well as uncovers a major memory leak.
</LI>
<LI>
The search programs
<B>blastn</B>, <B>tblastn</B>, and <B>tblastx</B>
may crash due to a segmentation fault
when a (nucleotide) database sequence
contains one or more ambiguity codes.
For <B>blastn</B>, a word length, W, less than 11 is also required
to elicit the bug.
This bug may appear sporadically and may disappear/reappear when using
the identical query and command line parameters,
if the database has been reprocessed with <B>pressdb</B>.
The chance of encountering the bug increases with:
shorter neighborhood word lengths;
lower values for the ungapped alignment score cutoff, S2;
and longer queries.
As a stop-gap measure,
the bug can be avoided entirely by making the FASTA-format database file
inaccessible to the search programs after processing by <B>pressdb</B>;
however, any ambiguity codes in the database sequences will not appear
in the search results and their absence during the search may yield
incorrect alignment scores.
</LI>
<LI>
Sometimes crashes occur when only the top or bottom strand
of the query sequence is being compared.
</LI>
<LI>
Occasional floating point exceptions may arise.
</LI>
<LI>
If the search program is interrupted during the preliminary filtering
of the query sequence,
a temporary file (usually created in the /var/tmp or /tmp directory)
used in the filtering process is not removed.
Over time, these files can accumulate, fill up the associated disk partition,
and thus make it impossible for any further searches involving filtering to proceed.
Other programs which use these partitions for temporary files
may have their operation impaired as well;
the "vi" editor is one such program.
</LI>
</UL>

<P>
The above mentioned bugs are not applicable to the licensed version of WU&nbsp;BLAST&nbsp;2.0,
but the licensed version does have some characteristics worth mentioning
that could trip up or confuse even the most knowledgeable of BLAST users.
Any unexpected behavior might be construed as a bug, so the following
information is provided to help avoid the unexpected.
If you should encounter problems or confusing areas
other than those described below,
or if you have questions or suggestions,
please send them to
<script language="JavaScript" type="text/javascript">
<!--
	var first = 'ma';
	var second = 'il';
	var third = 'to:';
	var address = '&#87;&#97;&#114;&#114;&#101;&#110;&#32;&#71;&#105;&#115;&#104;&#32;&#60;&#103;&#105;&#115;&#104;';
	var domain = '&#119;&#97;&#116;&#115;&#111;&#110;&#46;&#119;&#117;&#115;&#116;&#108;&#46;&#101;&#100;&#117;&#62;&#63;&#115;&#117;&#98;&#106;&#101;&#99;&#116;&#61;&#66;&#76;&#65;&#83;&#84;&#32;&#81;&#117;&#101;&#115;&#116;&#105;&#111;&#110;&#115;&#32;&#38;&#32;&#83;&#117;&#103;&#103;&#101;&#115;&#116;&#105;&#111;&#110;&#115;';
	document.write('<a href="');
	document.write(first+second+third);
	document.write(address);
	document.write('&#64;');
	document.write(domain);
	document.write('" title="Click here to compose an e-mail message">');
	document.write('BLAST Questions & Suggestions.<\/a>');
// -->
</script>
</P>

<UL>
<LI>
The statistical significance of <I>gapped</I> alignment scores is computed using
values for <I>&lambda;,</I> <I>K</I> and <I>H</I>
looked up in precomputed tables.
(The values for <I>&lambda;,</I> <I>K</I> and <I>H</I> used to assess the
significance of <I>ungapped</I> alignment scores are still computed at run time,
as is practical).
Values are chosen from the tables
based on the scoring matrix and gap penalties being used.
Precomputed values are not available for
all scoring matrix and gap penalty combinations, however;
and the precomputed values may not be well-suited
to an unusual residue composition of the query sequence.
When precomputed values are unavailable,
the programs issue a WARNING and proceed to evaluate gapped alignment scores
using values of <I>&lambda;,</I> <I>K</I> and <I>H</I>
that are known to be <I>in</I>correct:
the values computed for ungapped alignments.
In such cases, the reported significance estimates may be <I>highly</I> inaccurate
and will be biased towards being overly significant.
If the user knows more accurate values for their situation,
the <B>gapK</B>, <B>gapL</B> and <B>gapH</B> command line options
should be used to set them.</LI>

<LI>
Precomputed values for <I>&lambda;</I>, <I>K</I> and <I>H</I>
are available for <TT>BLASTN</TT> searches
with the following match,mismatch (M,N) scoring systems,
using gap penalties {Q,R}:
<PRE>
    "+1,-3", {3,3} {3,2} {3,1}
    "+1,-2", {2,2} {2,1} {1,1}
    "+3,-5", {10,5} {6,3} {5,5}
    "+4,-5", {10,5}
    "+1,-1", {3,1} {2,1}
    "+5,-4", {20,10} {10,10}
    "+5,-11", {22,22} {22,11} {12,2} {11,11}
</PRE>
and for the Purine-Pyrimidine scoring matrix:
<PRE>
pupy = 
	{ 20, 10}
	{ 10, 10}
</PRE>
</LI>
<LI>
Precomputed values for <I>&lambda;</I>, <I>K</I> and <I>H</I>
are available for protein-level searches
with the following scoring matrix and
gap penalty combinations (or gap penalty ranges for R) {Q, R}:
<PRE>
blosum50 =
    { 16,  1-4}
    { 15,  1-4,6,8}
    { 14,  1-5,8}
    { 13,  1-5,8}
    { 12,  2-5,7}
    { 11,  2-4,6,8}
    { 10,  2-6,8}
    { 9,  3-5,7}
    { 8,  4-8}
    { 7,  6,7}

blosum55 =
    { 16,  1-4}
    { 15,  1-4,5,6,8}
    { 14,  1-5,7}
    { 13,  2-5,8}
    { 12,  2-5,8}
    { 11,  2-6,8}
    { 10,  3-6,9}
    { 9,  3-5,7}
    { 8,  4-8}
    { 7,  7}

blosum62 =
    { 12,  1-3}
    { 11,  1-3}
    { 10,  1-4}
    { 9,  1-5}
    { 8,  2-7}
    { 7,  2-6}
    { 6,  3-5}
    { 5,  5}

blosum80 =
    { 12, 2-12}
    { 11, 2-11}
    { 10, 2-10}
    {  9,  3-9}
    {  8,  4-8}
    {  7,  5-7}

pam40 =
    { 12,  1,2,6}
    { 11,  1,2,7}
    { 10,  1-3,7}
    { 9,  1-3,6}
    { 8,  1-4}
    { 7,  1-4}
    { 6,  2-5}
    { 5,  2-5}
    { 4,  3,4}

pam120 =
    { 12,  1,2,4}
    { 11,  1-3}
    { 10,  1-3,5}
    { 9,  1-3,5}
    { 8,  1-4,6}
    { 7,  2-4,6}
    { 6,  2-5}
    { 5,  3-5}

pam250 =
    { 16,  1-4}
    { 15,  1-5}
    { 14,  1-6}
    { 13,  1-6}
    { 12,  2-7}
    { 11,  2-7}
    { 10,  3-8}
    { 9,  3-7}
    { 8,  5-7}
    { 7,  7}
</PRE>
</LI>



<LI>
Selecting an alternative scoring matrix does <I>not</I> alter
the gap penalties (<B>Q</B> and <B>R</B>)
from their default values.
This can not only result in alignments with undesirable gap characteristics,
but
depending on the scoring matrix chosen,
this can unwittingly create a situation in which
the programs do not have precomputed values
for <I>&lambda;,</I> <I>K</I> and <I>H.</I>
As described earlier, a WARNING message will be displayed when precomputed
values are not available;
nevertheless, the statistics will be unreliable.</LI>

<LI>
The <B>hspsepqmax</B>, <B>gapsepqmax</B>, etc. parameters are measures of distance
in residues along the sequences in the specific form in which they are
being compared.
For instance, in a <TT>BLASTX</TT> search (conceptually translated nt.
query sequence compared against a protein sequence database),
hspsepqmax refers to a distance measured in amino acid residues, not the underlying
nucleotides in the query.</LI>

<LI>
In <B>blastn</B> output,
if an alignment contains no gaps, the mid-line displayed
between the aligned query and subject sequences will contain vertical bars ("|")
to indicate identities;
but if there are one or more gaps in the alignment, the midline will contain
nucleotide residue codes to indicate positions of identity.
This was originally intended to be a feature, enabling gapped and ungapped
alignments to be readily discerned by eye, but it has caused some
confusion and, therefore, was removed from the licensed version.
</LI>

<LI>
The gap penalty parameters <B>Q</B> and <B>R</B> of WU&nbsp;BLAST
have similar but important differences in interpretation
from the parameters <B>G</B> and <B>E</B> of NCBI&nbsp;Gapped BLAST.
While the two extension penalties <B>R</B> (WU&nbsp;BLAST) and <B>E</B> (NCBI&nbsp;BLAST)
are analogous,
<B>Q</B> (WU&nbsp;BLAST) is analogous to the <I>sum</I> of <B>G</B> and <B>E</B>
with NCBI&nbsp;BLAST.
In other words, where <B>Q</B> is the total penalty for a gap of length 1,
NCBI Gapped BLAST computes this penalty as <B>G + E</B>.
</LI>

<LI>
ASN.1 formatted output is broken in all releases
of WU&nbsp;BLAST&nbsp;2.0.</LI>
</UL>

<HR>

<A NAME="Platforms"></A>
<H3>Supported Platforms</H3>
<P>
The computing platforms currently supported by
BLAST&nbsp;2.0 (licensed version only) include the following:
</P>

<UL>
<LI>Apple Mac&nbsp;OS X 10.1 and 10.2 ("Jaguar") for both PowerPC G3 and G4</LI>
<LI>Compaq Tru64 UNIX V4.0F for Alpha (upwardly compatible with Tru64 V5.X)</LI>
<LI>FreeBSD 4.5 for Intel i686 (PentiumPro/II/III)</LI>
<LI>Hewlett-Packard HP-UX 11 for HP PA-RISC and Intel IA-64 (Itanium)</LI>
<LI>IBM AIX 5.1 for Power3 and Power4</LI>
<LI>Linux kernel version 2.2 for i586 (original Pentium) and i686</LI>
<LI>Linux kernel version 2.4 for i686, i786 (Pentium4), and IA-64</LI>
<LI>SGI IRIX 6.5 for MIPS R5000, R10000 and R12000</LI>
<LI>Sun Solaris 8 for SPARC and Solaris 9 for Intel i686</LI>
</UL>
<B>The list of supported platforms is subject to change without notice.</B>
<BR>
Multiple processors
(multithreading or parallel processing)
are effectively and efficiently
supported by WU&nbsp;BLAST on all of the above platforms.
WU&nbsp;BLAST&nbsp;2.0 also supports large files
(files greater than 2&nbsp;GB in size)
when the underlying operating system and file system support large files.
<P>
Under Mac&nbsp;OS&nbsp;X,
WU&nbsp;BLAST is the only BLAST that
runs faster on multiple G4 processors.
Other BLAST implementations either don't use multiple G4 processors or
&mdash; as is the case with Apple's modified version of NCBI BLAST
&mdash; actually run <I>slower</I> when two processors are used.
Unlike other BLASTs, WU&nbsp;BLAST
won't crash or hang your system when the use of multiple CPUs is attempted
and it yields the most accurate results.
WU&nbsp;BLAST also does not utilize any G4-specific instructions
for peak performance,
so you can even run it on an iBook.
You don't have to scrap your old G3 in order to run the fastest BLAST &mdash;
WU&nbsp;BLAST runs on the same hardware that Mac&nbsp;OS&nbsp;X runs on
&mdash;
and it has done so since even before Mac&nbsp;OS&nbsp;X was publicly released &mdash;
but it will run faster on a G4 and again up to twice as fast on a dual G4.
</P>

<P>
Hewlett-Packard HP-UX and IBM AIX operating systems may need to be patched
for error-free support of large files over NFS.
If a large file/NFS problem does exist with your HP or IBM system,
it should immediately reveal itself
when an attempt is made to search a large-file database over NFS:
the search will simply fail to run and the application will exit non-zero.
If necessary,
simply contact your vendor for the patch.
Both companies promptly addressed this issue over a year ago,
in the first half of 2001.
</P>

<P>
Please note: while WU&nbsp;BLAST version 2.0a19 binaries dated February 1998 are freely available
<A HREF="http://blast.wustl.edu/blast/executables" TARGET="_blank">here</A>
for some platforms,
newer, full-featured binaries for the above platforms are only available upon
<A HREF="#Licensing">licensing</A>.
While containing several bugs,
some of the more prominent features <I>missing</I>
from WU&nbsp;BLAST&nbsp;2.0a19
but supported by the licensed version are:
</P>

<UL>
<LI>support for databases of virtually unlimited size &mdash; without splitting &mdash;
as well as the other unique and powerful features of the XDF database format,
such as complete identifier indexing and rapid appending;</LI>
<LI>support for so-called <I>virtual databases</I>
comprised of multiple real databases that are combined at run time;</LI>
<LI>significantly improved speed;</LI>
<LI>64-bit virtual addressing (on supporting platforms);</LI>
<LI>more efficient 32-bit virtual addressing on many 64-bit platforms;</LI>
<LI>two-hit BLAST option in all search modes;</LI>
<LI>many of the new command line options described above,
which improve the speed, sensitivity and selectivity of the software.</LI>
<LI>uniform support for large files (>2 GB) by all programs in the suite,
for any and all input and output files;
</LI>
</UL>

<HR>
<A NAME="Installation"></A>
<H3>Installation</H3>
To install WU-BLAST 2.0,
the first step is to download the UNIX tar archive of executables appropriate
for your computing platform from
<A HREF="http://blast.wustl.edu/blast/executables" TARGET="WU-BLAST 2.0 Executables">here</A>.
Scoring matrix files are included in each package.
<P>
Unpack the archive in a new, empty directory.
For convenience, precompiled versions of the low-complexity sequence filters
(<I>e.g.,</I> <A HREF="http://blast.wustl.edu/pub/seg">seg</A>,
<A HREF="http://blast.wustl.edu/pub/xnu">xnu</A>,
and
<A HREF="http://blast.wustl.edu/pub/dust">dust</A>)
are recent additions to the archive (see the <I>filter/</I> subdirectory created),
along with the sequence redundancy remover
<A HREF="http://blast.wustl.edu/pub/nrdb">nrdb</A>.
</P>

<P>
The executable programs from the tar archive may be moved as necessary
into any directory listed in the PATH environment variable,
whether this means adding the newly created directory to the PATH
or moving the executables
into an existing directory already listed in the PATH.
(Lots of information about interrogating and setting environment variables
&mdash;
and about the PATH environment variable itself
&mdash;
can be found in Google and other
search engines using keywords &ldquot;PATH environment variable&rdquo;).
If the software is installed in a directory that was already listed
in the PATH,
it may be necessary to exit the currently open shell and open a new one
in order for the shell to recognize the existence
of the newly installed programs.
</P>

<P>
Note that the files blastp, blastn, blastx, tblastn and tblastx
are actually <B><I>soft links</I></B> to a single copy of the <B>blasta</B> executable containing
the integrated capabilities of all 5 search methods.
These links should be retained
if the <B>blasta</B> file is moved to another directory.
If desired, the links can be renamed, as long as the original names appear as substrings
within the new names.
Alphabetic case is unimportant.
For instance, a link named wu-blastp will still invoke blasta in its blastp operational mode.
</P>

<P>
<I>A Note to Mac OS X Users</I>
<BR>
This software is invoked via a CLI (command line interface).
Programs will need to be invoked either via the Terminal application
(located in the /Applications/Utilities folder) or from within a script
or other application provided by a third party.
The latter may be clickable to execute, but the programs bundled in this
package are not intended to be double-clicked.
</P>

<P>
<I>A Note About File Permissions and File Copying</I>
<BR>
The WU-BLAST 2.0 package is copyrighted and only available under license.
To help ensure users of the software
do not unintentionally copy or distribute it outside your locale,
all copies of binary files are recommeneded to be maintained on your computer(s)
with execute-only permissions.
As delivered in the software archives from Washington University,
execute-only permissions have already been set,
but if the binary files should be copied by you,
these permissions may become altered and thus allow other users
to then copy the software in an unrestricted manner.
Restoration of execute-only permissions to a binary file
can be accomplished by running the command "chmod a-rw,a+x <I>filename</I>",
where <I>filename</I> is the name of the binary file.
</P>

<P>
<I>If you already had BLAST 1.4 installed (with BLAST-able databases),
the installation steps for WU-BLAST 2.0 are now complete.
If you did not have BLAST 1.4 installed,
or if you wish to see some of the flexibility afforded by using the <B>xdformat</B> program,
read on...</I>
</P>

<P>
Unpacking the tar archive creates a <B>matrix/</B> subdirectory containing scoring
matrix files.
Wherever this directory ultimately resides,
the <B>WUBLASTMAT</B> (BLASTMAT) environment variable should be set to point there.
In the absence of this environment variable being set,
the programs first look for scoring matrix files in any <I>matrix/</I> subdirectory
of the search program's directory and then in <CODE>/usr/ncbi/blast/matrix</CODE>.
</P>

<P>
Low-complexity sequence filters or masking programs &mdash;
<I>e.g.</I>,
<A HREF="http://blast.wustl.edu/pub/seg" TARGET="seg">seg</A>,
<A HREF="http://blast.wustl.edu/pub/xnu" TARGET="xnu">xnu</A>
and
<A HREF="http://blast.wustl.edu/pub/dust" TARGET="dust">dust</A> &mdash;
are now <I>included</I> in the tar archives described here.
While these filter programs are not required for running the search programs,
they can enormously reduce the amount of garbage output produced,
memory used, and search time taken.
Hence, it is highly recommended that these programs be made available to users.
If you wish to build (compile and link) the programs yourself,
their source code is posted off the
<A HREF="http://blast.wustl.edu">WU BLAST Archives</A> home page.
Whatever directory you install the filter programs in,
the <B>WUBLASTFILTER</B> (BLASTFILTER) environment variable should be set to point there.
In the absence of this environment variable being set,
the programs look for masking programs in any <I>filter/</I> subdirectory of the search program's
directory and in <CODE>/usr/ncbi/blast/filter</CODE>.
</P>

<P>
<I>Note:</I> unlike the NCBI BLAST 2.0 search programs,
the WU BLAST search programs do <I>not</I> employ sequence filtering by default.
This behavior may change, however, as the executable filter programs are now
distributed with the search programs.
In case the search programs are updated on your system without warning
and you wish to guarantee for an automated analysis pipeline
that no filtering is performed,
specify <CODE>filter=none</CODE> on the command line.
</P>

<P>
The databases themselves are missing from the tar archives, however!
Once the source databases have been downloaded from any of many Internet sites,
the database files are typically uncompressed and processed into FASTA format.
Included in the tar archives are several utility programs for converting
textual database files:
</P>

<UL>
<LI><B>gb2fasta</B> converts the nucleotide sequences in
<A HREF="http://ncbi.nlm.nih.gov/genbank" TARGET="GenBank Flat Files">GenBank flat files</A>
into FASTA format.
</LI>
<LI><B>gt2fasta</B> converts the CDS translations in
<A HREF="http://ncbi.nlm.nih.gov/genbank" TARGET="GenBank Flat Files">GenBank flat files</A>
into FASTA format.
</LI>
<LI><B>sp2fasta</B> converts
<A HREF="http://ftp.ebi.ac.uk/pub/databases/embl" TARGET="EMBL Flat Files">EMBL</A>
or SWISS-PROT flat files into FASTA format.
</LI>
</UL>
The NCBI software
<A HREF="http://ncbi.nlm.nih.gov/toolbox" TARGET="NCBI Toolbox">Toolbox</A> also contains some relevant parsers.
One of these is
<B>asn2fast</B>, which converts both nucleotide and peptide sequences in
<A HREF="http://ncbi.nlm.nih.gov/ncbi-asn1" TARGET="GenBank ASN.1">GenBank ASN.1 format</A>
into FASTA format files.
The <B>asn2ff</B> parser,
which converts GenBank ASN.1 data into other flat file formats,
may also come in handy, especially if you are inclined to parse GenBank
into FASTA using your own routines
or to using the gb2fasta and gt2fasta programs mentioned above.
<P>
All of the above parsers can read from standard input (sometimes signified by a single dash, "-"),
so their input files can be maintained on disk in compressed format and dynamically zcat-ed or gunzip-ed
directly into the parsers, thus saving the time and storage required for the uncompressed data.
Because a dash is often used to signify the start of each command line option,
if a dash is needed to specify standard input for the required input filename argument,
some of these programs require that a double-dash (--)
be specified on the command line before the single-dash.
This double-dash signifies the end of the command line options and the
start of the required arguments.
</P>

<P>
Once the databases are in FASTA format,
the <B>xdformat</B>, <B>setdb</B> and <B>pressdb</B> programs
are used to convert them into a blastable format.
Concise usage instructions for these programs can be obtained by invoking them
with no command line arguments.
By default,
each of these programs produces 3 output files whose names are derived from the name
of the FASTA input file.
The 3 output files have distinct filename extensions and
together comprise the blastable database.
When nucleotide sequences contain ambiguity codes (<I>e.g.</I>, EST data
which often contain many Ns),
a database formatted by <B>pressdb</B> will require that
the original FASTA input file remain accessible
to the search programs (<I>e.g.</I> BLASTA), if they are to retrieve the
ambiguity codes for matching sequences that contain them;
however, the FASTA input need <I>not</I> be accessible to the search programs
when databases are formatted by <B>xdformat</B>,
as this program stores full ambiguity information within the output files it produces.
More information about the blastable database file formats is available
<A HREF="http://blast.wustl.edu/blast/dbfmts.html">here</A>.
</P>

<P>
While formatting the database,
the <B>xdformat</B> program can optionally (<B>-I</B> option)
index the sequence identifiers
for later identifier-based retrieval with the <B>xdget</B> program.
XDF databases that were formatted without an identifier index
can have an index created <I>post hoc</I> by <B>xdformat</B>
with its <B>-X</B> option.
It may be of interest to note
for the purposes of their maintenance
that <B>xdformat</B> and <B>xdget</B> are actually one-and-the-same
program file,
merely invoked under the two different names to obtain the two
different behaviors.
This helps ensure that the index created with <B>xdformat</B>
will be compatible with <B>xdget</B>.
See the file "FAQ-Indexing.html" for more details on identifier indexing.
</P>

<P>
For compatibility with previous BLAST installations,
the <B>xdformat</B> program can function in a <B>setdb</B>- and <B>pressdb</B>-compatibility mode,
wherein its behavior is similar to that of setdb and pressdb.
In its compatibility mode, a similar command line structure is used
and the output files produced have the same names as those produced by setdb and pressdb.
Compatibility mode is invoked
when xdformat is renamed or has links pointing to it named <B>setdb</B> and <B>pressdb</B>.
While the output files produced in compatibility mode have the same names as those
produced by the corresponding
setdb/pressdb program, the <I>content</I> of the files produced is always <B>XDF</B>.
Versions of the <B>BLASTA</B> search program dated <B>1999-12-14</B> and later are
able to recognize and interpret these files properly.
</P>

<P>
Which instance of a database does BLAST search
if <B>xdformat</B>, setdb and pressdb
have been used to produce multiple databases each having the same base filename?
The database having the native filename extensions
of <B>xdformat</B> is chosen preferentially.
</P>

<P>
Using the -t option to <B>xdformat</B>,
a descriptive name or title can be assigned to a database that will appear in BLAST search output.
The title of an existing database can be changed after its creation,
by appending an empty FASTA database and specifying the -t option with the desired new title.
For example,
</P>

<PRE>

     xdformat -n -a olddb -t "A New Title" /dev/null

</PRE>

<P>
The blastable database files can be placed anywhere,
but for convenience the <B>BLASTDB</B> environment variable
should include their directory location.
If the <B>BLASTDB</B> environment variable is not set,
the programs look for databases by default
in <CODE>/usr/ncbi/blast/db</CODE> and in the current working
directory.
If the old pressdb program (instead of xdformat)
is used to create the blastable database,
the associated nucleotide sequence FASTA file must be located
in the same directory as the three output files from pressdb,
if the BLAST search programs are to find the FASTA file.
It may sometimes be useful to maintain the FASTA files
in a separate directory &mdash; even on another disk partition &mdash; and
provide UNIX soft links in the <B>BLASTDB</B> directory that point to the
real location of the FASTA files.
In addition, on systems where NCBI BLAST will not be in use,
blastable databases can be maintained in multiple directories listed
in the <B>BLASTDB</B> environment variable,
with each directory name delimited from the next by a colon (<B>:</B>),
just as directory names are often delimited
in the PATH environment variable.
</P>

<P>
On multi-processor computer systems,
the search programs will employ
as many CPUs as are installed;
when more than about 4 CPUs are used,
this default behavior cause efficiency of hardware utilization
to be quite low,
compared to running individual single-threaded BLAST jobs
on each CPU.
Memory use also increases linearly with the number of CPUs or threads
employed.
One way to govern the number of processors employed is
to wrap the search programs in a shell script that sets a lower
number of CPUs via the <B>cpus=#</B> command line option.
Another, simpler approach to changing the default number of CPUs
for all users follows below,
for implementation by BLAST system managers possessing
"root" or "SuperUser" privileges.
</P>

<P>
Licensed distributions of WU&nbsp;BLAST&nbsp;2.0
include a sample file named <B>sysblast.sample</B>,
that illustrates the system-wide configuration parameters
that can be established to govern the execution of <TT>BLAST</TT> jobs
and, thereby, provide a more productive, trouble-free level of service.
When the sysblast file is installed under the name <B>/etc/sysblast</B>,
all <TT>BLAST</TT> jobs executed on a given computer system can be made
subject to the parameters:
</P>

<UL>
<LI><B>cpusmax</B>=<I>&lt;n&gt;</I>:  a hard limit on the number of CPUs or threads employed
by each <TT>BLAST</TT> job;
it is possible to prohibit <TT>BLAST</TT> searches entirely
on a given computer by configuring a negative value for <B>cpusmax</B>;
</LI>
<LI><B>cpus</B>=<I>&lt;n&gt;</I>:  the default number of CPUs or threads employed per <TT>BLAST</TT> job;
</LI>
<LI><B>nice</B>=<I>&lt;n&gt;</I>:  a "nice" value for altering the priority of <TT>BLAST</TT> processes;
		As is standard for UNIX operating systems:
	<UL>
		<LI>positive nice values correspond to <I>lower</I> priority
		</LI>
		<LI>only the root user can run at negative nice values (higher priority);
		</LI>
		<LI>any nice value set in /etc/sysblast is added to the current nice value of a <TT>BLAST</TT> process.</LI>
	</UL>
</LI>
<LI><B>memmax</B>=<I>&lt;n&gt;</I>:  the maximum amount of memory that may be
allocated by any single <TT>BLAST</TT> job.
	The interpretation and recommended usage of <B>memmax</B> are:
	<UL>
	<LI>
	<I>memmax</I> is expressed in units of bytes,
	with optional modifiers <B>k</B> (KiB), <B>m</B> (MiB),
	and <B>g</B> (GiB);
	</LI>
	<LI>
	It is almost certainly a <B><I>bad idea</I></B> to set memmax to a value that is
	greater than the actual amount of memory (silicon RAM) installed
	in the computer;
	</LI>
	<LI>
	If <I>memmax=0</I>, the effective limit is "unlimited", or the natural
	upper limit for a process executing under the given operating system;
	</LI>
	<LI>
	Values of <I>memmax&nbsp;&lt;&nbsp;0</I> are ignored, in which case
	the standard UNIX <B>datasize</B> resource limit set by the
	user's command shell governs <TT>BLAST</TT> memory usage instead;
	</LI>
	</UL>
</LI>
</UL>

<P>
The sysblast file is only effective when installed in the /etc directory.
The /etc directory resides locally to any given computer system,
so parameter settings can be tailored to each computer,
even if the <TT>BLAST</TT> software is maintained on a shared disk partition.
The /etc directory should only be writable by "root".
Unlike the shell script wrapper approach described above, 
the limits set in /etc/sysblast typically can not be circumvented
by normal (non-root) users of a computer system.
See the comments included in the sample sysblast file for further details.
</P>

<P>
For further information on <TT>BLAST</TT> software
installation, the out-dated
<A HREF="http://blast.wustl.edu/blast-1.4/blast1.pdf" TARGET="_blank">manual</A>
page for the BLAST version 1.4 (ungapped) search programs
sometimes is useful,
as it provides a description
of procedures and parameters that have not changed in BLAST&nbps;2.0.
</P>


<HR>

<A NAME="Licensing"></A>
<H3>Licensing</H3>
<P>
Site licenses for the full-featured BLAST&nbsp;2.0 are available free
for academic and nonprofit use;
commercial licenses are available from Washington University for a fee.
Academic and nonprofit licenses are typically arranged through the
institutions' respective offices of technology transfer.
Upon obtaining written permission from Washington University,
licensees are welcome to install the software for public BLAST services.
Washington University seeks additional licensees
for commercial development and marketing
and invites interested parties to submit proposals.
</P>

<P>
Please address all e-mail requests for licensing information
and limited evaluation copies to
<script language="JavaScript" type="text/javascript">
<!--
	var first = 'ma';
	var second = 'il';
	var third = 'to:';
	var address = '&#83;&#111;&#102;&#116;&#119;&#97;&#114;&#101;&#32;&#76;&#105;&#99;&#101;&#110;&#115;&#105;&#110;&#103;&#32;&#60;&#108;&#105;&#99;&#101;&#110;&#115;&#105;&#110;&#103;';
	var domain = '&#98;&#108;&#97;&#115;&#116;&#46;&#119;&#117;&#115;&#116;&#108;&#46;&#101;&#100;&#117;&#62;&#63;&#115;&#117;&#98;&#106;&#101;&#99;&#116;&#61;&#66;&#76;&#65;&#83;&#84;&#32;&#76;&#105;&#99;&#101;&#110;&#115;&#105;&#110;&#103;';
	document.write('<a href="');
	document.write(first+second+third);
	document.write(address);
	document.write('&#64;');
	document.write(domain);
	document.write('" title="Click here to compose an e-mail request">');
	document.write('BLAST Software Licensing.<\/a>');
// -->
</script>
</P>

<P>
Be sure to include the <B>name and address of your company or institution</B>
and the <B>name and e-mail address of your lab head</B> (if not you).
Washington University typically negotiates site licenses for BLAST,
so only one license needs to be executed per institution.
If your institution already has a license,
you will be informed of this upon inquiring, and your lab head
will be provided with download instructions.
If a license for your institution does not exist,
you may be provided with a draft of the license agreement,
which will need to be signed by authorized representatives of both institutions.
<B>If you do not receive a response within the next business day,
please re-send your message and indicate it is a repeat request.</B>
Please note that responses during holiday periods may be slower than usual.
</P>

<HR>
<A NAME="Citing"></A>
<H3>
Citing BLAST</H3>
<P>
Citations or acknowledgements of WU&nbsp;BLAST usage are greatly appreciated,
as are any personal accounts of how the software is being used
that you might wish to share.
When URLs are acceptable, please cite with:
</P>

<PRE>
   Gish, W. (1996-2003) http://blast.wustl.edu
</PRE>

<P>
When URLs are not acceptable, please use:
</P>

<PRE>
   Gish, W., personal communication.
</PRE>

<P>
The WU BLAST search program may also be referred
to by the name <B>BLASTA</B>.
I know of no other program (BLAST-related or otherwise)
going by this name.
</P>

<P>
In scientific communications,
it is typically important to report the
program name, as well as the specific version used.
In the case of WU&nbsp;BLAST or BLASTA,
the version is a combination of the "2.0" moniker and the release date.
The release date can be found on the first line of output,
and it is the first date displayed.
For example, consider this introductory line of output:
</P>

<PRE>
  BLASTN 2.0MP-WashU [<B>02-Apr-2002</B>] [sol8-ultra-ILP32F64 2002-04-03T01:25:46]
</PRE>

<P>
The software release date is April 2, 2002,
whereas the compilation or build date
of the Solaris 8 binary
was April 3rd at 1:25 AM.
</P>

<HR>
<A NAME="History"></A>
<H3>
Historical Notes</H3>
<P>
Historical notes and additional citation information
for some earlier versions of NCBI and WU BLAST include:
</P>

<UL>
<LI>The first description of the classical ungapped BLAST
algorithm was published by
<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=2231712&amp;dopt=Abstract" TARGET="_blank">
Altschul <I>et al.</I> (1990)</A>.
This paper focuses on <TT>BLASTP</TT> and <TT>BLASTN</TT>,
and makes mention of <TT>TBLASTN</TT>.
(It is perhaps of interest that
<TT>TBLASTN</TT> hadn't been written
at the time the manuscript was submitted,
but it was available by the time of publication).
</LI>
<LI>
The NCBI
<A HREF="http://blast.wustl.edu/network/experimental" TARGET="_blank">
<I>Experimental BLAST Network Service</I></A>
was opened to the public in December 1989,
providing Internet access to the latest, parallelized search programs
and sequence databases updated on a daily basis.
Around the same time, the
"<A HREF="http://blast.wustl.edu/pub/nrdb" TARGET="_blank">nr</A>"
(quasi-non-redundant) databases were established (W. Gish, unpublished).
The experimental service was ultimately discontinued
more than a decade later in March 2000.
At the request of NCBI upper management,
a report on the experimental service was never published
and remains W. Gish (unpublished).
Awareness of the service spread by word-of-mouth,
much as is the case with WU&nbsp;BLAST.
</LI>
<LI>
<TT>BLASTX</TT> first appeared in
<A HREF="http://blast.wustl.edu/blast-1.1" TARGET="_blank">
BLAST&nbsp;1.1</A> in July 1990,
and was later described by
<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=8485583&amp;dopt=Abstract" TARGET="_blank">
Gish and States (1993)</A>.
The <B>BLAST3</B>
program (<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=2196570&amp;dopt=Abstract" TARGET="_blank">
Altschul and Lipman, 1990</A>) was also folded into this release
and parallelized.
The use of Poisson statistics, as suggested by
<A HREF="http://www.ncbi.nlm.nih.gov/htbin-post/Entrez/query?uid=2315319&amp;form=6&amp;db=m&amp;Dopt=b" TARGET="_blank">
Karlin and Altschul (1990)</A>
to evaluate the joint probability of multiple HSPs,
was also first featured in BLAST&nbsp;1.1.
</LI>
<LI>
<TT>BLASTC</TT>,
a version of <TT>BLASTX</TT> that considered codon usage information
in addition to sequence similarity
(<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=8790452&amp;dopt=Abstract" TARGET="_blank">States and Gish, 1994</A>),
only appeared in the
<A HREF="http://blast.wustl.edu/blast-1.3" TARGET="_blank">
BLAST&nbsp;1.3</A>
distribution.
The BLAST&nbsp;1.3 distribution was also the last to include the <B>BLAST3</B>
program.
</LI>
<LI>
The first version of BLAST to use
<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=8390686&amp;dopt=Abstract" TARGET="_blank">
Karlin and Altschul (1993)</A>
"Sum" statistics
to evaluate the joint probability of multiple HSPs
was
<A HREF="http://blast.wustl.edu/blast-1.4" TARGET="_blank">
BLAST&nbsp;1.4</A>
(W. Gish, unpublished).
</LI>
<LI>
<TT>TBLASTX</TT> first appeared in BLAST version 1.4
and remains attributable to W. Gish (unpublished).
</LI>
<LI>
The first release of <B>WU&nbsp;BLAST</B> was version&nbsp;1.4,
which was virtually identical to NCBI BLAST&nbsp;1.4,
save for a few bug fixes.
The <B>WU&nbsp;BLAST&nbsp;Archives</B> (<A HREF="http://blast.wustl.edu"><I>http://blast.wustl.edu</I></A>)
first appeared on the Internet in 1995,
to provide continued support for the work begun at the NCBI,
as well as to provide a central location where BLAST-related software,
information, and earlier software versions could be obtained.
</LI>
<LI>
Starting in late 1994,
Stephen Altschul and I engaged in a collaboration to provide
support for my conjecture that fixed <I>&lambda;</I>, <I>K</I> and <I>H</I> values,
along with Sum statistics,
could be practically applied to the
evaluation of locally optimal <I>gapped</I> alignment scores.
This work eventually appeared in
<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=8743700&amp;dopt=Abstract" TARGET="_blank">
Altschul and Gish (1996)</A>
and provides much of the foundation for today's WU&nbsp;BLAST&nbsp;2.0
and NCBI&nbsp;blastall.
</LI>
<LI>
The first complete implementation of gapped BLAST
(<TT>BLASTP</TT>, <TT>BLASTN</TT>, <TT>BLASTX</TT>, <TT>TBLASTN</TT> and <TT>TBLASTX</TT>)
with statistical significance estimates (both Poisson and Sum)
was publicly released as <B>WU&nbsp;BLAST version 2.0d1</B>
(W. Gish, unpublished),
in time for presentation
at the Cold Spring Harbor Genome Mapping and Sequencing conference
in May 1996.
</LI>
<LI>
The NCBI published its BLAST version 2, or <B>Gapped BLAST</B>,
including a description of the <B>2-hit</B> BLAST and <B>PSI-BLAST</B> algorithms,
in
<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=9254694&amp;dopt=Abstract" TARGET="_blank">
Altschul <I>et al.</I> (1997)</A>,
in September 1997.
</LI>
<LI>
The NCBI published a description of <B>PHI-BLAST</B> in
<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=9705509&amp;dopt=Abstract" TARGET="_blank">
Zhang <I>et al.</I> 1998</A>.
</LI>
</UL>

<HR>

<A NAME="References"></A>
<H3>References</H3>

<P>
<A NAME="Altschul_and_Gish_1996">
Altschul, SF, and W Gish (1996).
</A>
<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=8743700&amp;dopt=Abstract" TARGET="_blank">
Local alignment statistics.
</A>
ed. R.&nbsp;Doolittle. Methods in Enzymology&nbsp;<B>266</B>:460-80.
</P>

<P>
<A NAME="Altschul_and_Lipman_1990">
Altschul, SF, and DJ Lipman (1990).
</A>
<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=2196570&amp;dopt=Abstract" TARGET="_blank">
Protein database searches for multiple alignments.
</A>
Proc.&nbsp;Natl.&nbsp;Acad.&nbsp;Sci.&nbsp;USA&nbsp;<B>87</B>:5509-13.
</P>

<P>
<A NAME="Altschul_et_al_1990">
Altschul, SF, Gish, W, Miller, W, Myers, EW, and DJ Lipman (1990).
</A>
<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=2231712&amp;dopt=Abstract" TARGET="_blank">
Basic local alignment search tool.
</A>
J.&nbsp;of&nbsp;Mol.&nbsp;Biol.&nbsp;<B>215</B>:403-10.
</P>

<P>
<A NAME="Altschul_et_al_1997">
Altschul, SF, Madden, TL, Schaffer, AA, Zhang, J, Zhang, Z, Miller, W, and DJ Lipman (1997).
</A>
<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=9254694&amp;dopt=Abstract" TARGET="_blank">
Gapped BLAST and PSI-BLAST: a new generation of protein database search programs.
</A>
Nucleic&nbsp;Acids&nbsp;Res.&nbsp;<B>25</B>(17):3389-402.
</P>

<P>
<A NAME="Claverie_and_States_1993">
Claverie, JM, and DJ States (1993).</A>
Information enhancement methods for large scale sequence analysis.
Computers in Chemistry&nbsp;<B>17</B>:191-201.
</P>

<P>
<A NAME="Gish_and_States_1993">
Gish, W, and DJ States (1993).
</A>
<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=8485583&amp;dopt=Abstract" TARGET="_blank">
Identification of protein coding regions by database similarity search.
</A>
Nature Genetics&nbsp;<B>3</B>:266-72.
</P>

<P>
<A NAME="Hancock_and_Armstrong_1994">
Hancock, JM, and JS Armstrong (1994).</A>
<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=7514951&amp;dopt=Abstract" TARGET="_blank">
SIMPLE34: an improved and enhanced implementation
for VAX and Sun computers of the SIMPLE algorithm
for analysis of clustered repetitive motifs in nucleotide sequences.</A>
Comput.&nbsp;Appl.&nbsp;Biosci.&nbsp;<B>10</B>:67-70.
</P>

<P>
<A NAME="Karlin_and_Altschul_1990">
Karlin, S, and SF Altschul (1990).
</A>
<A HREF="http://www.ncbi.nlm.nih.gov/htbin-post/Entrez/query?uid=2315319&amp;form=6&amp;db=m&amp;Dopt=b" TARGET="_blank">
Methods for assessing the statistical significance of molecular sequence
features by using general scoring schemes.</A>
Proc.&nbsp;Natl.&nbsp;Acad.&nbsp;Sci.&nbsp;USA&nbsp;<B>87</B>:2264-8.
</P>

<P>
<A NAME="Karlin_and_Altschul_1993">
Karlin, S, and SF Altschul (1993).</A>
<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=8390686&amp;dopt=Abstract" TARGET="_blank">
Applications and statistics for multiple high-scoring segments
in molecular sequences.
</A>
Proc. Natl. Acad. Sci.&nbsp;<B>90</B>:5873-7.
</P>

<P>
<A NAME="Smith_Waterman_1981">
Smith, TF, and MS Waterman (1981).</A>
<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=7265238&amp;dopt=Abstract" TARGET="_blank">
Identification of common molecular subsequences.
</A>
J.&nbsp;Mol.&nbsp;Biol.&nbsp;<B>147</B>:195-7.
</P>

<P>
<A NAME="States_Gish_1994">
States, DJ, and W Gish (1994).</A>
<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=8790452&amp;dopt=Abstract" TARGET="_blank">
Combined use of sequence similarity and codon bias for coding region
identification.
</A>
J.&nbsp;Comp.&nbsp;Biol.&nbsp;<B>1</B>:39-50.
</P>

<P>
<A NAME="Wootton_and_Federhen_1993">
Wootton, JC, and S Federhen (1993).</A>
Statistics of local complexity in amino acid sequences
and sequence databases.
Computers in&nbsp;Chemistry&nbsp;<B>17</B>:149-63.
</P>

<P>
<A NAME="Wootton_and_Federhen_1996">
Wootton, JC, and S Federhen (1996).
</A>
<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=8743706&amp;dopt=Abstract" TARGET="_blank">
Analysis of compositionally biased regions in sequence databases.</A>
ed. R.&nbsp;Doolittle. Methods in Enzymology&nbsp;<B>266</B>:554-71.
</P>

<P>
<A NAME="Zhang_et_al_1998">
Zhang, Z, Schaffer, AA, Miller, W, Madden, TL, Lipman, DJ, Koonin, EV,
and SF Altschul (1998).
</A>
<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;list_uids=9705509&amp;dopt=Abstract" TARGET="_blank">
Protein sequence similarity searches using patterns as seeds.
</A>
Nucleic Acids Res.&nbsp;<B>26</B>:3986-90.
</P>

<HR>

Return to the
<A HREF="http://blast.wustl.edu"><EM>WU&nbsp;BLAST&nbsp;Archives</EM></A> home page

</BODY>
</HTML>

